{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 40 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "/media/Home_Raid1_Voyager/kian/anaconda3/lib/python3.7/site-packages/gtfparse/read_gtf.py:154: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  features=features)\n",
      "/media/Home_Raid1_Voyager/kian/anaconda3/lib/python3.7/site-packages/gtfparse/read_gtf.py:154: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  features=features)\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_name', 'exon_number', 'exon_id']\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['transcript' 'exon' 'CDS' 'start_codon' 'stop_codon' '3UTR' '5UTR']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>NR_046018.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>NR_046018.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>3</td>\n",
       "      <td>NR_046018.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname      source     feature  start    end  score strand frame  gene_id  \\\n",
       "0    chr1  ncbiRefSeq  transcript  11874  14409    NaN      +     0  DDX11L1   \n",
       "1    chr1  ncbiRefSeq        exon  11874  12227    NaN      +     0  DDX11L1   \n",
       "2    chr1  ncbiRefSeq        exon  12613  12721    NaN      +     0  DDX11L1   \n",
       "3    chr1  ncbiRefSeq        exon  13221  14409    NaN      +     0  DDX11L1   \n",
       "4    chr1  ncbiRefSeq  transcript  14362  29370    NaN      -     0   WASH7P   \n",
       "\n",
       "  transcript_id gene_name exon_number        exon_id  \n",
       "0   NR_046018.2   DDX11L1                             \n",
       "1   NR_046018.2   DDX11L1           1  NR_046018.2.1  \n",
       "2   NR_046018.2   DDX11L1           2  NR_046018.2.2  \n",
       "3   NR_046018.2   DDX11L1           3  NR_046018.2.3  \n",
       "4   NR_024540.1    WASH7P                             "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gtfparse, os, pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "savingdir = os.getcwd()\n",
    "refseq_file = os.path.join('/media/Scratch_SSD_Voyager/Kian/genomes/hg38/hg38.ncbiRefSeq.gtf')\n",
    "refseq_df = gtfparse.read_gtf(refseq_file)\n",
    "\n",
    "\n",
    "annot_file = os.path.join('/media/Scratch_SSD_Voyager/Kian/genomes/hg38/annotations/gencode.v41.annotation.gtf')\n",
    "annot_df = gtfparse.read_gtf(annot_file)\n",
    "\n",
    "print(refseq_df['feature'].unique())\n",
    "refseq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM']\n"
     ]
    }
   ],
   "source": [
    "chrms = ['chr{}'.format(n) for n in [*np.arange(1, 23), 'X', 'Y', 'M']]\n",
    "print(chrms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2846745</th>\n",
       "      <td>chr3</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>155554076</td>\n",
       "      <td>155554196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PLCH1</td>\n",
       "      <td>NM_014996.3</td>\n",
       "      <td>PLCH1</td>\n",
       "      <td>15</td>\n",
       "      <td>NM_014996.3.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551721</th>\n",
       "      <td>chr17</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4072608</td>\n",
       "      <td>4072756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>NM_015113.4</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>31</td>\n",
       "      <td>NM_015113.4.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3617633</th>\n",
       "      <td>chr7</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>108159463</td>\n",
       "      <td>108159541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NRCAM</td>\n",
       "      <td>NM_001371158.1</td>\n",
       "      <td>NRCAM</td>\n",
       "      <td>2</td>\n",
       "      <td>NM_001371158.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503753</th>\n",
       "      <td>chr16</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>70120461</td>\n",
       "      <td>70120719</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>PDPR</td>\n",
       "      <td>NM_017990.5</td>\n",
       "      <td>PDPR</td>\n",
       "      <td>3</td>\n",
       "      <td>NM_017990.5.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469771</th>\n",
       "      <td>chr10</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>68033965</td>\n",
       "      <td>68034186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>HERC4</td>\n",
       "      <td>NM_001278185.2</td>\n",
       "      <td>HERC4</td>\n",
       "      <td>19</td>\n",
       "      <td>NM_001278185.2.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3434193</th>\n",
       "      <td>chr6_GL000250v2_alt</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>3172158</td>\n",
       "      <td>3172376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>SNHG32</td>\n",
       "      <td>NR_160953.1_2</td>\n",
       "      <td>SNHG32</td>\n",
       "      <td>4</td>\n",
       "      <td>NR_160953.1_2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2969296</th>\n",
       "      <td>chr4</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>82907785</td>\n",
       "      <td>82907935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>THAP9</td>\n",
       "      <td>NM_024672.6</td>\n",
       "      <td>THAP9</td>\n",
       "      <td>4</td>\n",
       "      <td>NM_024672.6.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2664043</th>\n",
       "      <td>chr3</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>33373847</td>\n",
       "      <td>33373921</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>FBXL2</td>\n",
       "      <td>NR_146133.2</td>\n",
       "      <td>FBXL2</td>\n",
       "      <td>10</td>\n",
       "      <td>NR_146133.2.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428750</th>\n",
       "      <td>chr16</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16202001</td>\n",
       "      <td>16202178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ABCC6</td>\n",
       "      <td>NM_001351800.1</td>\n",
       "      <td>ABCC6</td>\n",
       "      <td>23</td>\n",
       "      <td>NM_001351800.1.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1680827</th>\n",
       "      <td>chr17</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>59047683</td>\n",
       "      <td>59047819</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TRIM37</td>\n",
       "      <td>NM_001320987.3</td>\n",
       "      <td>TRIM37</td>\n",
       "      <td>9</td>\n",
       "      <td>NM_001320987.3.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     seqname      source feature      start        end  score  \\\n",
       "2846745                 chr3  ncbiRefSeq    exon  155554076  155554196    NaN   \n",
       "1551721                chr17  ncbiRefSeq    exon    4072608    4072756    NaN   \n",
       "3617633                 chr7  ncbiRefSeq    exon  108159463  108159541    NaN   \n",
       "1503753                chr16  ncbiRefSeq    exon   70120461   70120719    NaN   \n",
       "469771                 chr10  ncbiRefSeq    exon   68033965   68034186    NaN   \n",
       "3434193  chr6_GL000250v2_alt  ncbiRefSeq    exon    3172158    3172376    NaN   \n",
       "2969296                 chr4  ncbiRefSeq    exon   82907785   82907935    NaN   \n",
       "2664043                 chr3  ncbiRefSeq    exon   33373847   33373921    NaN   \n",
       "1428750                chr16  ncbiRefSeq    exon   16202001   16202178    NaN   \n",
       "1680827                chr17  ncbiRefSeq    exon   59047683   59047819    NaN   \n",
       "\n",
       "        strand frame gene_id   transcript_id gene_name exon_number  \\\n",
       "2846745      -     0   PLCH1     NM_014996.3     PLCH1          15   \n",
       "1551721      -     0   ZZEF1     NM_015113.4     ZZEF1          31   \n",
       "3617633      -     0   NRCAM  NM_001371158.1     NRCAM           2   \n",
       "1503753      +     0    PDPR     NM_017990.5      PDPR           3   \n",
       "469771       -     0   HERC4  NM_001278185.2     HERC4          19   \n",
       "3434193      +     0  SNHG32   NR_160953.1_2    SNHG32           4   \n",
       "2969296      +     0   THAP9     NM_024672.6     THAP9           4   \n",
       "2664043      +     0   FBXL2     NR_146133.2     FBXL2          10   \n",
       "1428750      -     0   ABCC6  NM_001351800.1     ABCC6          23   \n",
       "1680827      -     0  TRIM37  NM_001320987.3    TRIM37           9   \n",
       "\n",
       "                   exon_id  \n",
       "2846745     NM_014996.3.15  \n",
       "1551721     NM_015113.4.31  \n",
       "3617633   NM_001371158.1.2  \n",
       "1503753      NM_017990.5.3  \n",
       "469771   NM_001278185.2.19  \n",
       "3434193    NR_160953.1_2.4  \n",
       "2969296      NM_024672.6.4  \n",
       "2664043     NR_146133.2.10  \n",
       "1428750  NM_001351800.1.23  \n",
       "1680827   NM_001320987.3.9  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keeping only exons (which include UTRs) filter the predicted genes\n",
    "refseq_df = refseq_df.loc[(refseq_df['feature'] == 'exon') & ((refseq_df['transcript_id'].str.startswith('N')) | (refseq_df['seqname'] == 'chrM'))]\n",
    "refseq_df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get immunoglobulin genes from Gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>409795</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>87338511</td>\n",
       "      <td>87338559</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000233999.3</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKV3OR2-268-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>mRNA_end_NF,cds_end_NF,basic,Ensembl_canonical...</td>\n",
       "      <td>OTTHUMT00000330418.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003567656.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000474297.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409798</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>87338735</td>\n",
       "      <td>87339035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000233999.3</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKV3OR2-268-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>mRNA_end_NF,cds_end_NF,basic,Ensembl_canonical...</td>\n",
       "      <td>OTTHUMT00000330418.3</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003560657.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000474297.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411003</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88857161</td>\n",
       "      <td>88857683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000211592.8</td>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKC-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>mRNA_start_NF,cds_start_NF,basic,Ensembl_canon...</td>\n",
       "      <td>OTTHUMT00000323482.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507664.2</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000374777.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411009</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88860568</td>\n",
       "      <td>88860605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000211593.2</td>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ5-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>mRNA_start_NF,mRNA_end_NF,cds_start_NF,cds_end...</td>\n",
       "      <td>OTTHUMT00000323474.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507665.2</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000420166.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411013</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88860886</td>\n",
       "      <td>88860922</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000211594.2</td>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ4-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>mRNA_start_NF,mRNA_end_NF,cds_start_NF,cds_end...</td>\n",
       "      <td>OTTHUMT00000323495.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002047359.1</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000417097.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  source feature     start       end  score strand frame  \\\n",
       "409795    chr2  HAVANA    exon  87338511  87338559    NaN      +     0   \n",
       "409798    chr2  HAVANA    exon  87338735  87339035    NaN      +     0   \n",
       "411003    chr2  HAVANA    exon  88857161  88857683    NaN      -     0   \n",
       "411009    chr2  HAVANA    exon  88860568  88860605    NaN      -     0   \n",
       "411013    chr2  HAVANA    exon  88860886  88860922    NaN      -     0   \n",
       "\n",
       "                  gene_id  gene_type  ...   transcript_name  \\\n",
       "409795  ENSG00000233999.3  IG_V_gene  ...  IGKV3OR2-268-201   \n",
       "409798  ENSG00000233999.3  IG_V_gene  ...  IGKV3OR2-268-201   \n",
       "411003  ENSG00000211592.8  IG_C_gene  ...          IGKC-201   \n",
       "411009  ENSG00000211593.2  IG_J_gene  ...         IGKJ5-201   \n",
       "411013  ENSG00000211594.2  IG_J_gene  ...         IGKJ4-201   \n",
       "\n",
       "       transcript_support_level  \\\n",
       "409795                       NA   \n",
       "409798                       NA   \n",
       "411003                       NA   \n",
       "411009                       NA   \n",
       "411013                       NA   \n",
       "\n",
       "                                                      tag  \\\n",
       "409795  mRNA_end_NF,cds_end_NF,basic,Ensembl_canonical...   \n",
       "409798  mRNA_end_NF,cds_end_NF,basic,Ensembl_canonical...   \n",
       "411003  mRNA_start_NF,cds_start_NF,basic,Ensembl_canon...   \n",
       "411009  mRNA_start_NF,mRNA_end_NF,cds_start_NF,cds_end...   \n",
       "411013  mRNA_start_NF,mRNA_end_NF,cds_start_NF,cds_end...   \n",
       "\n",
       "           havana_transcript exon_number            exon_id ont  \\\n",
       "409795  OTTHUMT00000330418.3           1  ENSE00003567656.1       \n",
       "409798  OTTHUMT00000330418.3           2  ENSE00003560657.1       \n",
       "411003  OTTHUMT00000323482.1           1  ENSE00001507664.2       \n",
       "411009  OTTHUMT00000323474.3           1  ENSE00001507665.2       \n",
       "411013  OTTHUMT00000323495.2           1  ENSE00002047359.1       \n",
       "\n",
       "               protein_id ccdsid artif_dupl  \n",
       "409795  ENSP00000474297.1                    \n",
       "409798  ENSP00000474297.1                    \n",
       "411003  ENSP00000374777.2                    \n",
       "411009  ENSP00000420166.1                    \n",
       "411013  ENSP00000417097.1                    \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df = annot_df.loc[annot_df['gene_type'].isin(['IG_C_gene', \"IG_D_gene\",\n",
    "                                                   \"IG_J_gene\", \"IG_V_gene\", 'TR_C_gene', \n",
    "                                                   'TR_J_gene', 'TR_V_gene', 'TR_D_gene'])]\n",
    "annot_df = annot_df.loc[(annot_df['feature'] == 'exon')]\n",
    "\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>409795</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>87338511</td>\n",
       "      <td>87338559</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGKV3OR2-268</td>\n",
       "      <td>ENST00000421835.2</td>\n",
       "      <td>IGKV3OR2-268</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003567656.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409798</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>87338735</td>\n",
       "      <td>87339035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGKV3OR2-268</td>\n",
       "      <td>ENST00000421835.2</td>\n",
       "      <td>IGKV3OR2-268</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003560657.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411003</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88857161</td>\n",
       "      <td>88857683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>ENST00000390237.2</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507664.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411009</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88860568</td>\n",
       "      <td>88860605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>IGKJ5</td>\n",
       "      <td>ENST00000390238.2</td>\n",
       "      <td>IGKJ5</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507665.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411013</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>88860886</td>\n",
       "      <td>88860922</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>IGKJ4</td>\n",
       "      <td>ENST00000390239.2</td>\n",
       "      <td>IGKJ4</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002047359.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193212</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>22910574</td>\n",
       "      <td>22910606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGLJ4</td>\n",
       "      <td>ENST00000390326.2</td>\n",
       "      <td>IGLJ4</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507829.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193219</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>22914237</td>\n",
       "      <td>22914308</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGLJ5</td>\n",
       "      <td>ENST00000390327.2</td>\n",
       "      <td>IGLJ5</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507830.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193226</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>22918132</td>\n",
       "      <td>22918201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGLJ6</td>\n",
       "      <td>ENST00000390328.2</td>\n",
       "      <td>IGLJ6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507831.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193233</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>22921390</td>\n",
       "      <td>22921435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGLJ7</td>\n",
       "      <td>ENST00000390330.2</td>\n",
       "      <td>IGLJ7</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001507834.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193237</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>22922594</td>\n",
       "      <td>22923034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IGLC7</td>\n",
       "      <td>ENST00000390331.3</td>\n",
       "      <td>IGLC7</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001665273.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>766 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source feature     start       end  score strand frame  \\\n",
       "409795     chr2  HAVANA    exon  87338511  87338559    NaN      +     0   \n",
       "409798     chr2  HAVANA    exon  87338735  87339035    NaN      +     0   \n",
       "411003     chr2  HAVANA    exon  88857161  88857683    NaN      -     0   \n",
       "411009     chr2  HAVANA    exon  88860568  88860605    NaN      -     0   \n",
       "411013     chr2  HAVANA    exon  88860886  88860922    NaN      -     0   \n",
       "...         ...     ...     ...       ...       ...    ...    ...   ...   \n",
       "3193212   chr22  HAVANA    exon  22910574  22910606    NaN      +     0   \n",
       "3193219   chr22  HAVANA    exon  22914237  22914308    NaN      +     0   \n",
       "3193226   chr22  HAVANA    exon  22918132  22918201    NaN      +     0   \n",
       "3193233   chr22  HAVANA    exon  22921390  22921435    NaN      +     0   \n",
       "3193237   chr22  HAVANA    exon  22922594  22923034    NaN      +     0   \n",
       "\n",
       "              gene_id      transcript_id     gene_name exon_number  \\\n",
       "409795   IGKV3OR2-268  ENST00000421835.2  IGKV3OR2-268           1   \n",
       "409798   IGKV3OR2-268  ENST00000421835.2  IGKV3OR2-268           2   \n",
       "411003           IGKC  ENST00000390237.2          IGKC           1   \n",
       "411009          IGKJ5  ENST00000390238.2         IGKJ5           1   \n",
       "411013          IGKJ4  ENST00000390239.2         IGKJ4           1   \n",
       "...               ...                ...           ...         ...   \n",
       "3193212         IGLJ4  ENST00000390326.2         IGLJ4           1   \n",
       "3193219         IGLJ5  ENST00000390327.2         IGLJ5           1   \n",
       "3193226         IGLJ6  ENST00000390328.2         IGLJ6           1   \n",
       "3193233         IGLJ7  ENST00000390330.2         IGLJ7           1   \n",
       "3193237         IGLC7  ENST00000390331.3         IGLC7           1   \n",
       "\n",
       "                   exon_id  \n",
       "409795   ENSE00003567656.1  \n",
       "409798   ENSE00003560657.1  \n",
       "411003   ENSE00001507664.2  \n",
       "411009   ENSE00001507665.2  \n",
       "411013   ENSE00002047359.1  \n",
       "...                    ...  \n",
       "3193212  ENSE00001507829.2  \n",
       "3193219  ENSE00001507830.2  \n",
       "3193226  ENSE00001507831.2  \n",
       "3193233  ENSE00001507834.2  \n",
       "3193237  ENSE00001665273.3  \n",
       "\n",
       "[766 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>NR_046018.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>NR_046018.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>3</td>\n",
       "      <td>NR_046018.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>NR_024540.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>NR_024540.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780871</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5842</td>\n",
       "      <td>5933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NR_001537.1_2</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>2</td>\n",
       "      <td>NR_001537.1_2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780872</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8560</td>\n",
       "      <td>8721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NR_001537.1_2</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>3</td>\n",
       "      <td>NR_001537.1_2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780873</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10331</td>\n",
       "      <td>10429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NR_001537.1_2</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>4</td>\n",
       "      <td>NR_001537.1_2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780874</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11025</td>\n",
       "      <td>11186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NR_001537.1_2</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>5</td>\n",
       "      <td>NR_001537.1_2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780875</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15645</td>\n",
       "      <td>15725</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NR_001537.1_2</td>\n",
       "      <td>TTTY13</td>\n",
       "      <td>6</td>\n",
       "      <td>NR_001537.1_2.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>780876 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    seqname      source feature  start    end  score strand  \\\n",
       "0                      chr1  ncbiRefSeq    exon  11874  12227    NaN      +   \n",
       "1                      chr1  ncbiRefSeq    exon  12613  12721    NaN      +   \n",
       "2                      chr1  ncbiRefSeq    exon  13221  14409    NaN      +   \n",
       "3                      chr1  ncbiRefSeq    exon  14362  14829    NaN      -   \n",
       "4                      chr1  ncbiRefSeq    exon  14970  15038    NaN      -   \n",
       "...                     ...         ...     ...    ...    ...    ...    ...   \n",
       "780871  chrY_KZ208924v1_fix  ncbiRefSeq    exon   5842   5933    NaN      -   \n",
       "780872  chrY_KZ208924v1_fix  ncbiRefSeq    exon   8560   8721    NaN      -   \n",
       "780873  chrY_KZ208924v1_fix  ncbiRefSeq    exon  10331  10429    NaN      -   \n",
       "780874  chrY_KZ208924v1_fix  ncbiRefSeq    exon  11025  11186    NaN      -   \n",
       "780875  chrY_KZ208924v1_fix  ncbiRefSeq    exon  15645  15725    NaN      -   \n",
       "\n",
       "       frame  gene_id  transcript_id gene_name exon_number          exon_id  \n",
       "0          0  DDX11L1    NR_046018.2   DDX11L1           1    NR_046018.2.1  \n",
       "1          0  DDX11L1    NR_046018.2   DDX11L1           2    NR_046018.2.2  \n",
       "2          0  DDX11L1    NR_046018.2   DDX11L1           3    NR_046018.2.3  \n",
       "3          0   WASH7P    NR_024540.1    WASH7P           1    NR_024540.1.1  \n",
       "4          0   WASH7P    NR_024540.1    WASH7P           2    NR_024540.1.2  \n",
       "...      ...      ...            ...       ...         ...              ...  \n",
       "780871     0   TTTY13  NR_001537.1_2    TTTY13           2  NR_001537.1_2.2  \n",
       "780872     0   TTTY13  NR_001537.1_2    TTTY13           3  NR_001537.1_2.3  \n",
       "780873     0   TTTY13  NR_001537.1_2    TTTY13           4  NR_001537.1_2.4  \n",
       "780874     0   TTTY13  NR_001537.1_2    TTTY13           5  NR_001537.1_2.5  \n",
       "780875     0   TTTY13  NR_001537.1_2    TTTY13           6  NR_001537.1_2.6  \n",
       "\n",
       "[780876 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gen2refseq = pd.DataFrame(annot_df[['seqname', 'source', 'feature', 'start', 'end', 'score', \n",
    "                           'strand', 'frame', 'gene_id', 'transcript_id', 'gene_name',\n",
    "                           'exon_number', 'exon_id']])\n",
    "gen2refseq.loc[:, 'gene_id'] = gen2refseq['gene_name'].to_numpy()\n",
    "display(gen2refseq)\n",
    "refseq_df = pd.concat([gen2refseq, refseq_df], ignore_index=True).sort_values(by=['seqname', 'start'], ignore_index=True)\n",
    "display(refseq_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making target files for the constitutive exons per chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findConstitutiveExons2(exonStarts, exonEnds, cutoff = 0.75):\n",
    "    \"\"\"\n",
    "        finding the constitutive exons from the multiple isoforms of a gene\n",
    "        exonStarts: array containing the start coordinate of exons\n",
    "        exonEnds: array containing the end coordinate of exons\n",
    "    \"\"\"\n",
    "    geneStart = np.min(exonStarts)\n",
    "    geneEnd = np.max(exonEnds)\n",
    "    geneLen = geneEnd - geneStart + 1\n",
    "    collapse = np.zeros(geneLen)\n",
    "\n",
    "    for i, (start, end) in enumerate(zip(exonStarts, exonEnds)):\n",
    "        collapse[(start - geneStart) : (end - geneStart + 1)] += 1\n",
    "\n",
    "    n_iso = collapse.max()\n",
    "    overlaps = np.where(collapse >= n_iso * cutoff)[0]\n",
    "\n",
    "    noDirectionBounds = np.where((overlaps - np.append(overlaps[1:], overlaps[-1]+3)) < -1)[0]\n",
    "    cExonStarts, cExonEnds = [], []\n",
    "    for i, bound in enumerate(noDirectionBounds):\n",
    "        if i == 0:\n",
    "            cExonStarts.append(geneStart + overlaps[0])\n",
    "        else:\n",
    "            cExonStarts.append(geneStart + overlaps[noDirectionBounds[i-1] + 1])\n",
    "\n",
    "        cExonEnds.append(geneStart + overlaps[bound])\n",
    "    return cExonStarts , cExonEnds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ref_chr = refseq_df.loc[refseq_df['seqname'].str.split('_').str[0] == ('chr1')]\n",
    "\n",
    "# ref_chr[['seqname', 'gene_name']].groupby(by = ['seqname', 'gene_name']).size().reset_index().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matt's version of ppDesigner gives probes within an exon range expanded by flankingLen. So I need to shrink each exon from both sides by flankingLen so that I get probes within the exon \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing targets for chr1\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "Preparing targets for chr2\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "Preparing targets for chr3\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "Preparing targets for chr4\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "Preparing targets for chr5\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "Preparing targets for chr6\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "Preparing targets for chr7\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "Preparing targets for chr8\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "Preparing targets for chr9\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "Preparing targets for chr10\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "Preparing targets for chr11\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "Preparing targets for chr12\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "Preparing targets for chr13\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "Preparing targets for chr14\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "Preparing targets for chr15\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "Preparing targets for chr16\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "Preparing targets for chr17\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "Preparing targets for chr18\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "Preparing targets for chr19\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "Preparing targets for chr20\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "Preparing targets for chr21\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "Preparing targets for chr22\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "Preparing targets for chrX\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "Preparing targets for chrY\n",
      "0\n",
      "100\n",
      "Preparing targets for chrM\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "flankingLen = 0 \n",
    "# iterate over chromosomes\n",
    "for chrm in chrms:\n",
    "    print(\"Preparing targets for {}\".format(chrm))\n",
    "    ppd_targets_chr = []\n",
    "    ref_chr = refseq_df.loc[refseq_df['seqname'].str.split('_').str[0] == (chrm)]\n",
    "        \n",
    "    \"\"\"Because some genes can be located on the chromosome patches \n",
    "        (e.g. chr1_KZ208906v1_fix), we have to iterate over chromosome patch-gene pairs\"\"\"\n",
    "    genes_chr = ref_chr.groupby(by = ['seqname', 'gene_name']).size().reset_index()\n",
    "    \n",
    "    # iterate over chromosome-gene pairs in the chromosome \n",
    "    for i, (chrmp, gene, _) in genes_chr.iterrows():\n",
    "        ref_gene = ref_chr.loc[(ref_chr['seqname'] == chrmp) & (ref_chr['gene_id'] == gene)]\n",
    "\n",
    "        # find the constitutive exons from the isoforms of the gene\n",
    "        cExonStarts, cExonEnds = findConstitutiveExons2(ref_gene['start'], ref_gene['end'], cutoff = 1/3)\n",
    "\n",
    "        \n",
    "        # Shrinking the exons by flankingLen from both sides and remove those\n",
    "        cExonStarts, cExonEnds = np.array(cExonStarts), np.array(cExonEnds)\n",
    "        cExonStarts += flankingLen\n",
    "        cExonEnds -= flankingLen\n",
    "        tooSmalls = cExonEnds <= cExonStarts\n",
    "        cExonStarts = cExonStarts[~tooSmalls]\n",
    "        cExonEnds = cExonEnds[~tooSmalls]\n",
    "        \n",
    "        # rearraging the exons for ppDesigner\n",
    "        n_exon = len(cExonStarts)\n",
    "        if n_exon > 0:\n",
    "            gene_ppd = {'gene_id' : n_exon * [gene], \n",
    "                        'seqname' : n_exon * [chrmp],\n",
    "                        'start' : cExonStarts,\n",
    "                        'end' : cExonEnds,\n",
    "                        'strand' : n_exon * [ref_gene['strand'].iloc[0]]}\n",
    "            gene_ppd = pd.DataFrame(gene_ppd)\n",
    "            gene_ppd['gene_id'] = gene_ppd['gene_id'] + '_' + gene_ppd['seqname'] + '_' + np.arange(1, n_exon + 1).astype(str)\n",
    "            ppd_targets_chr.append(gene_ppd)\n",
    "#         if i> 3:\n",
    "#             break\n",
    "        if i % 100 == 0:\n",
    "            print(i)\n",
    "    ppd_targets_chr = pd.concat(ppd_targets_chr, ignore_index=True)\n",
    "    \n",
    "    # saving the targets per chromosome in designated folder\n",
    "    if not os.path.isdir(os.path.join(savingdir, chrm)):\n",
    "        os.mkdir(os.path.join(savingdir, chrm))\n",
    "    ppd_targets_chr.to_csv(os.path.join(savingdir, chrm, '{}.target'.format(chrm)), \n",
    "                           sep = '\\t', header = False, index = False)\n",
    "    if os.path.exists(os.path.join(savingdir, chrm, 'ppd_targets_{}.txt'.format(chrm))):\n",
    "        os.remove(os.path.join(savingdir, chrm, 'ppd_targets_{}.txt'.format(chrm)))\n",
    "#   \n",
    "#     break\n",
    "# ref_chr\n",
    "# ref_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[150113839, 150117618, 150118747, 150119467, 150120012, 150120888, 150121204, 150121880, 150123042, 150124250, 150124727, 150125445, 150126520, 150129757, 150130539, 150131979, 150132750, 150133586, 150133881, 150134750, 150135555, 150137008, 150155397]\n",
      "[150115946, 150117850, 150118852, 150119566, 150120123, 150121010, 150121322, 150122040, 150123201, 150124360, 150124831, 150125577, 150126614, 150129968, 150130662, 150132094, 150132942, 150133760, 150134008, 150135016, 150135878, 150137053, 150155845]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, '#Exon Overlap')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA24AAACqCAYAAADVwBxYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAA9hUlEQVR4nO3dd7xcdZ3/8dcnvUECSWgJSUCKUgSWiCAoxVVAdEEEV3cFsWEXG4jl5+ruuuK6i4gFRcVeFgRRAaUHRFEIBESalFASWhISUslN7v38/pgzc8/MnfI9bWbu3Pfz8UjunTPzLefM95x7vufbzN0RERERERGR7jWq0xkQERERERGR5lRxExERERER6XKquImIiIiIiHQ5VdxERERERES6nCpuIiIiIiIiXU4VNxERERERkS6nipuIyAhiZmvNbOeUYReY2Tuj3//VzK7KN3fdxcw+ZWbfzTnO1MdfRERGNlXcRER6kJk9YmYboopC+d8O7j7F3R/OGr+7/9TdX51HXusxs38xs4VRvp80s9+Z2SEFpneYmS2Jb3P3/3L3ckV1npm5mY1JEGelohuLM5fjX5POYWY2UPNdrzWzg/JMR0REOiv4D5CIiAw7r3P3azqdiaTM7KPAmcB7gCuBPuAo4Fjgpg5mrZs94e6zO50JEREpjlrcRERGkKjVaJfo9x+Y2TfM7HIzW2NmfzGzF8Q++yozu8/MnjOzrwMWe+8UM7upJt73mNkDZrYyitei90ab2f+a2XIzW2xmH2jUemVmU4F/B97v7pe4+zp33+Tuv3X306PPjDezc8zsiejfOWY2PnrvMDNbYmYfM7Nnota6t8Xif42Z3RPt71Iz+7iZTQZ+B+wQb500s8+Z2U+ioDdGP1eVW7Nq3q9qlTOzLwAvB74eff7r8eNvZgea2VNmNjoW/vVm9tfo91FmdqaZPWRmK8zsQjPbOsX3vXV0PF4XvZ5iZg+a2cnR65eZ2a3Rd3yrmb0sFnaBmf2Hmf0xOl5XmdmM6L0JZvaTKG+rorDbJs2fiIiEU8VNRGRkezPweWAr4EHgCwDRDfrFwGeAGcBDwMEt4not8BJgH+CNwJHR9ncBRwP7Av8AHNckjoOACcCvmnzm08CBUXz7AAdE+SzbDpgKzALeAXzDzLaK3vse8G533wLYC7jO3ddF+Xsi6so4xd2fqEnzFdHPadH7NzfJH+7+aeAPwAeiz3+g5v0/A+uAI2Kb/wX4WfT7hygdp0OBHYCVwDeapdkgH88Cbwe+Y2bbAF8B7nD3H0UVwcuBc4HpwNnA5WY2vSZPbwO2AcYBH4+2v5XSMd4xCvseYEPS/ImISDhV3EREetelUWvIKjO7tMFnLnH3W9x9M/BTSpUhgNcA97j7L919E3AO8FSL9M5y91Xu/hhwfSyuNwJfdfcl7r4SOKtJHNOB5VF+GvlX4N/d/Rl3X0ap4nlS7P1N0fub3P0KYC2we+y9PcxsS3df6e63t9inIv2cUsUZM9uC0jH/efTeu4FPR8dsI/A54IQmY+x2iH3X5X+TAdz9KuAi4FrgmChuot8fcPcfu/tmd/85cB/wuli833f3v7v7BuBCBr/TTZS+q13cvd/db3P31dkOh4iINKOKm4hI7zrO3adF/45r8Jl4ZWw9MCX6fQfg8fIb7u7x11niahHPCmBGi0lAdgAejb1+NNpWiaOm4hfPyxsoVZAeNbMbOjyBx8+A46NunscDt7t7eb/mAr8qV8KAe4F+oFF3xCdi33X537rY++dTamH8vruviLbVHkei17Nirxt9pz+mNP7wF1F31f82s7GB+y0iIimo4iYiIvU8SakbHADReLUdG3+8ZVzxiTOaxXMz8DzNu1M+QaliUzYn2taSu9/q7sdS6vp3KaVWJABvFbTOtnXApNjr7QLCxPNyD6WK0tFUd5OEUuX26JqK2AR3X9oin0NE4+i+DfwIeK9FYxwZehyhdCxbphG1Zn7e3fcAXkapm+zJSfMmIiLhVHETEZF6Lgf2NLPjo9avDzG0YhLqQuA0M5tlZtOATzT6oLs/B3yW0ri048xskpmNNbOjzey/o4/9HPiMmc2MxuJ9FvhJozjLzGycldafmxp1/1xNqRUL4GlgejQ5Sj3LgAEgvgbbHcArzGxOFO6TNWGervl8PT+jdGxfQak7Y9m3gC+Y2dwo7zPN7NgWcTXyqejn24H/AX4UVeauAHaz0tILY8zsn4E9gMtaRWhmh5vZ3lE8qyl1nexvEUxERDJQxU1ERIZw9+XAiZTGo60AdgX+mDK67wBXAX8FFlGqMGymwY2+u58NfJTShCPLKLU+fYBSCxnAfwILo/juAm6PtoU4CXjEzFZTmlDjLVGa91GqED4cdU+Md73E3ddTmrjlj9H7B7r71cD/Rfm4jaEVnq9SGpe20szObZCfnwOHUZokZXlN2N8AV5nZGuDPwEub7Fd8RszyvzeY2f6UjuXJ7t4PfIlSS+CZUZfJ1wIfo/QdnwG8tiYfjWwH/JJSpe1e4AYCKs8iIpKelYYtiIiItIeZHQ18y91ru+mJiIhIA2pxExGRQpnZRCutnzbGzGYB/0bz6f5FRESkhlrcRESkUGY2iVJXuhdSWuvrcuA0TR8vIiISThU3ERERERGRLqeukiIiIiIiIl1OFTcREREREZEuN6bTGYibMWOGz5s3r9PZEBERERER6YjbbrttubvPrN1eeMUtWpxzIbDU3V/b7LPz5s1j4cKFRWdJRERERESkK5nZo/W2t6Or5GmUFucUERERERGRFAqtuJnZbOAY4LtFpiMiIiIiItLLiu4qeQ5wBrBFwekU58Mfhjvu6HQu6tt3XzjnnPamWe94ZM1HOc4k8YSEiec1TR5r97UcR9J4G8WTRRFx1sbbKM40xzXrd1FUXN0u7fnWrmOUNp0s15G8zstW+QmNp9nfiDyPfZrrpIjICLXxAx/i0Wv/SP9A82XPZhzyUmZ+77w25Sq7wlrczOy1wDPufluLz51qZgvNbOGyZcuKyo6IiIiIiIwAq9b38ey6Pvr6B+h3b/jPGV7rWRe2ALeZfRE4CdgMTAC2BC5x97c0CjN//nzX5CQiIiIiIpLWwkee5YRv3cyP33EAL991yOSMXc/MbnP3+bXbC2txc/dPuvtsd58HvAm4rlmlTUREREREROrTAtwiIiIiItIzhlcHyHBtWYDb3RcAC9qRloiIiIiIiGGdzkKu1OImIiIiIiLS5VRxExERERGRnlHQ3Isdp4qbiIiIiIj0HOutnpKquImIiIiIiHQ7VdxERERERKRnFLVOdaep4iYiIiIiIj2nx3pKquImIiIiIiLS7VRxExERERGRntGbHSUDF+A2s+2AAygdh1vd/alCcyUiIiIiIpJFj/WVbNniZmbvBG4BjgdOAP5sZm8vOmMiIiIiIiJSEtLidjqwn7uvADCz6cCfgAuKzJiIiIiIiEhSPTqpZNAYtyXAmtjrNcDjxWRHREREREQkO+uxvpIhLW5Lgb+Y2a8pjXE7FrjFzD4K4O5nF5g/ERERERGRES+k4vZQ9K/s19HPLfLPjoiIiIiISHreo/NKtqy4ufvn25ERERERERGRvFhv9ZRsXXEzs5nAGcCewITydnc/osB8iYiIiIiISCRkcpKfAvcBOwGfBx4Bbm0VyMwmmNktZnanmd1tZmq5ExERERGRYvVmT8mgitt0d/8esMndb3D3twMHBoTbCBzh7vsA+wJHmVlIOBERERERkUx6rKdk0OQkm6KfT5rZMcATwOxWgdzdgbXRy7HRv2FV/3V3Ng90b5bHjg6pd+dvU/9A1evRZowalf7UGBhw+t0ZZcbowHjKYVql3T/gDCSMO25z/wAOQ8KXj8GYUYYFdKAux5MkTCvlfYPs30FcOa+N4iyfFwaMCSyD8XMpj7yW9z2vY9nNas+30PO+1feYl7TpxM+JJGUJqst++dwsXxMg3TkWj7MsdJ9qv6OyTl2jRUSkN4VU3P7TzKYCHwO+BmwJfCQkcjMbDdwG7AJ8w93/kjajnbBk5QZe/t/XdzobDX3x+L158wFz2prmD//0CP/2m7urts3eaiI3nH54qorR2o2befmXrmPl+k1sMX4M159+GDOmjG8aZuPmfl7+pet5Zs1Gtp48jj+ccTiTxw8tyk899zyv/N8FrOvrZ/K40Vz7scPYbuqEOjHWt+ixlbzx2zezqd8ZP2YUl3/oEHbZZgvOW/AQX/r9fQDsP3crLn7vy5rGc8viZ/mX7/y5UnHZb840fvW+g4PzUc9dS57jDef9ib7ohnGP7bfkitNenilOgGvueZpTf7yQAYe9Zm3JZR8cGueH/+8Ofn3HE5jBV9+0H/+0zw4t4/3ohXfyq0VLAdh5xmSu+/hhqfP4t6XPcfw3S/t+/H6zOPuf900dV7f70u/v47wFD1VtCznvf3nbEj5+0Z0ATJ88jj984nAmjQu53Cdz3X1P884flsrLDlMncOMZhwdVwH5315O872e3VxZIHT3KuOCUl3DobjNbhq0t+xPHjuaqj7yCd/5wIfc/XVpy9JSXzeNz/7Rn8H48tmI9R55zIxs29Vdt323bKVz1kUObhj376r9z7rUP1H3vM8e8iHe+fOfgfIiISD66t9klm5BZJS+Lfn0OODxJ5O7eD+xrZtOAX5nZXu7+t/hnzOxU4FSAOXPaWwlpZcuJY/n4q3frdDbqOueaB3hk+bq2p7t4+TrGjxnFB4/YBYC/LH6WPzywnM0DA4weNTpxfKvW97Fy/SZ2njGZh5evY/najS0rbus29vPMmo3M3GI8y9ZsZPXzm+pW3J5e/Tzr+vrZZ/ZU7lzyHMvWbExUcVu6agOb+p1X7bEtV9/zNE8+9zy7bLMFi5evZcr4MeyyzRQWB3wHjz+7ns0DzjsP2YmFj64MCtM6b+vp6x/g5IPmcs8Tq7n7idWZ4wR49Nn1lUrb4mX18/nwsnXMmz6JR1as59HAfXl42Vp2mjGZbbccz58ffjZTHpes3EBf/wCTx43m4Q6cA+308LK1zJgynlNeNheIzvsVrfe5fG149R7bctU9T7N6w+ZCKm6PriiVlwN22ppbFj/Lxs0DQRW3R1asxx1Oe+Wu9PUPcN6Ch3j82fVBaZbL/lsPmsvajf1cfPsSnlnzPA8vX8sBO23N48+uT1wunlr9PBs29XPC/rOZN30SADc+sJxFj61sGfbhZWvZatJY3nHITlXbv7ngoVzOdRERSa/XeuU0/EtuZl+jSYXV3T8Umoi7rzKzBcBRwN9q3jsfOB9g/vz5XVVBnjpxLB84YtdOZ6Our1//YMfSnjB2dOW42PUP8ocHlmeOc6eo4uYJSsDMKaWKW6swe80qVdzSrunxyhduw9X3PF21bYsJY3jx7Kk8GnATXXbyQfPY1D8QdOMd6s0HzOFXi5bmVnEr23P7qTzcoOIGMGf6ZB5ZEXajXTZv+iT2nj0tc8WtrF5lvde4w4wp4yrnW5LzfvQo44gXbsNV9zxd2Ho25XPvJfO24pbFyb/X9x72AtY8v5nzFjwUnMNymm9+6RyeWb2Ri29fUtn+knlb0be5frfF5nGWIn39frM4eJcZAKzv6w+quAFsNXnckL8VP/jToz37xFdERDqj2Z3PwiwRR8sIbIoqbROBfwS+lCVOqdaJmwJvUEtKUuGqFy7JA5FyHlqFqYyfSfmwpTZv5ddJ97X242mPVbM48roxDzm2jlcG+wbfbGfKVf3YzHq3K0RTATvd7oVHLSoR4eUh5xJRez4kPMkafTokmhFZBkVEulwe91rdqGHFzd1/GI1RO8vdT08R9/ZAOY5RwIWxbpeSkXVwnpz4TX1+LdDJIwpNO+uxqhfeSJ5js/yb7M2KmTGpVTbT7IZZvqW2k+dAO8XLTJJ9Ntq38GiWdNKGNWxI2HrbksUZ+z1BPPU+2mO9c0REhqVeuxY37Wvk7v1mtn+aiN39r8B+qXIlQZI+Vc4lzaLjT/GEu/HTcq95nTAvDV57k8+E5SP7USyiFW9IGg3ijG8PTbdei0jaSmxVXL36SC1Sb+/Cylzz13kZWg7DEqr7sdCwDbZl2cVK0jVFMijOJh/q8eIpIiJtFjJIZJGZ/Qa4CKgMenH3SwrLlXS1TrTw1A0TmJPMT1t67GlNmOY73Q2HpNeeouWtnccnbVLd8h3W67o5Ulp0RUR6UbuHDLRLSMVta2AFcERsmwOquHWQWWee5tamWRnbknWMW/l1wIk2dOxZg3F3lTxWvw7Pm1eHj167R93+zHJtIUyWN6K8lfpK5lUU4se20XdR3n8IvzCWx8XFxwumHnsY/TTy2+9u5T60+15Iq9bgMUo29iwprzkZ06STeLxkLMn49afUihuVi5Q7XFtZCzvW9VuPiz3yIiISotcewYUsB/C2dmRERrZ0LW6hcWcc49YtzQJt1GqPu+GIjMTvpVsN+28ixSRJIiIi7dZywR0z283MrjWzv0WvX2xmnyk+a9JMp1obap8uV1pQUuamHC5Jy53X3GWFjMUqvU6Xx9oWgfi+BsUZbyHI6YurHLdyq0NOhSEeb8PjytCZNlvG6zUtJFnymGKM3fDl1ZMBEXiO+GBrbOl1sQcqeXkYPO+txXk8JGzlmlHd6u6VbZb4elTv06Gzlta2ilaF7/nyKSLSnXr1+tt6pVT4DvBJYBNUJh15U5GZkpGn2Ba35HHnGX44aj2+p/MHZSR+L92qV8aD9cZeiIhIWa/dK4RU3Ca5+y012zYXkRkJ18luYtbg90xxpppePvBzGVt56qVTbj1LEmdpOvyclwOgmItSq31Ll2a2qdrzycPwUjsWMNF5H1sqorBZJTN2MbSqPCabkbJ2KQyPmoLTZGUwzvjSC+HqXiMIGwMrIiISKqTittzMXkB0H2dmJwBPFporCdINk5O02h4aX6LbrZoJTVp9MPMC3JX+ZtXpB8dTEyDPyUkapZE13maHLH6DnXRCiXpxJBXvXturs0Y1E9R9r+0LcFcSDpL3tWvosgQ5xRvYLVVERLpLr16aQ2aVfD9wPvBCM1sKLAb+tdBcSUudbGwoZAHuVC1ugcsBJI+6Jp3625I+2y+iday4BbjzXw4g77yOhBY3qC5niVqBaF/LfPpFtLN8j0MLVNr4vN5DngQR1bsWjJTyKSLS3XrrYhxScXvU3f/RzCYDo9x9TdGZkjCdaG1olGLanKR5Uh4aJusixENbynxo+imeyHfzAtxJ9y3tosmanCRMvd0L2uc2HZdG50jrcGHbgsNmPdeDW/Hr5adxYiOxRVhERIoT0lVysZmdDxwIrC04PzIs5P/0IlULTujnuuRhS7fko1focDbX1gW4UybWzUs6dG/ORESklaJnU+6UkIrb7sA1lLpMLjazr5vZIcVmS1rq0FTTjRfgTrkcQHla8AQLOg9ZgLvRQtHlPFZuDhNOEV4zYUH5tbtH05iHjjcalNcyDoPHOzwfYfFGsVqT41qe2j9BuqVjlnza+LpxUc5j70/+UF5UuqLJ91IVbvDjUTx556x+vEmWhyhLshRI6XOD3RoHJx6q3pZ2OYB69chW17ZGi8mHLt0gIiLF6eLng6m0rLi5+wZ3v9Ddjwf2A7YEbig8ZzKipGtxa9MYt4zhh6PhML179+dw5Oj4H8aMFaT4gxAREZFuFdLihpkdambfBG4HJgBvLDRX0lKvTU6SqiUmNEzGVp7Gk5Mka2kst9LlKc0kKaHxNtu1NAsblFrp8lwPIL+oupVTvZvJJiexlq3SWcXafVOFt8p/yeteVS248XzkfE1KH940wk1EpEN69frbcnISM1sM3AFcCJzu7uuKzpR0s+bdEtPGluQeaXA6+BafS7PUQJPwla6SKeNp9DoPefXlDjm28UpA6mUgMlxSK13l6N0LczNhU9S3eTmAhJWvvCuS2SciKqnfVbJ5JW4klkERkeGi157xhswquY+7ry48J5JItyzAnVuclTFuScIQFCbpTWWj8FXbSDbGqxxP3t9bEUsMQOt9S7MvuS8+3vH+ecUrLypdlmSfy+MQK/EUIMtsjFCbx8AZKWNjTwdHr8bHuKVQZz+SlNdG34vGuImISJ6adpU0s6OBy8xsuZktM7MbzOw1bcqbtNCJGXOKWoC7iDBZj0+jKezzWlYgi6ELcOcj6YLDwdO/F9Tq2KuzRmXVqcOStPKVX7o1r/OMO2HaIiLSBXr02tywxc3M3gW8GzgDWBhtng+cZWaz3f38ZhGb2Y7Aj4DtgAHgfHf/ai65lo4qpIUn+pl0zFhImKwz6zXtNpggzuHUPlSasbG7r3rD6XimleUbiB+for7JugtXJ5Cm1bTew4Ks+ze4H/HWzYyRks9DGhERSa/Xeuc06yr5EeAQd382tu26qBXuJqBpxQ3YDHzM3W83sy2A28zsane/J1uWBZJ308vLkOUAsvZDrNdHKSxEyxur0M81DO/VN6Xl16XeaxZ8Maie+jyfJ/TxG80iph1vPsYtPjlLWHyOV3WLy6J2OYheV9V9z8IebmSdNCSpwWn5w8Q/l7qbJbEu1rFxj2bgA8niatblsxR38zOi3rsjpXyKiEj7NOsqaTWVNgDcfUVIxO7+pLvfHv2+BrgXmJUql9JVCpnFMEWcoTdG2fOb1zR1+UTTFsMgr8NhyYJOaudTxl6spHTbLm3uH+DGvy/j4WVrO50VERHpkGYVt9Vmtk/txmjbmiSJmNk8SmvA/aXOe6ea2UIzW7hs2bIk0Y5oRbSyhKjt+lM7OUDi+GpaT0JiGXy63vwp/5C4Ex6w2ha7wTFu0WLShO131SdyaimtaiHI8a45fmwbH9dkC5CXwpRb6apn6EyVx1j3vC7vzZlZ7QLcoTNpZi37oVKPXYx9sHaB+9CgVRObENuWZgHuJq24IWPcGi0Zkme3iBsfWMbJF9zCid+6Ob9IRUR6VOgM5MNNs66SHwN+Y2bfB26j9CfoJcBbgbeEJmBmU4CLgQ/Xm50yGit3PsD8+fN7/DasNxQ5xi1RmOAWt2zy2t3h1ELUiy0oI007v8I0rXu5lrGMfznqdS3ttnNg3cZ+AFas6+twTkREpFMatri5+03AAdFnTgHeHv1+YPReS2Y2llKl7afufknm3EpFtwy27OgC3ASGSbHUQHXwRusBJM9v/lPiF3OD3rJF15LvS94LkHfLOVC06jFu6fa56CdiWb6JpK328fUVhy4HkP8i9yHqXyK0ALeIiOSr6Tpu7v408Nk0EVvpDuN7wL3ufnaaOKS5TsxYlvtyANHPJJWA+DpOzePO2ExeM2FB2gW4i5irvN7xLnWry3bXGjLxRzzp8MlJal+nPwjxbqIjcda+oCUb2nxcBrssBla+ck4/8wLcNRMRJYmrXUd65JV0EZH0enUis6bruGV0MHAScISZ3RH90xpwOelkObQGv2eKsxJRkuUAwsJkXYS4cYNbsifqec2qWJ2PYloYWlWkrfJfokhzLbe9djGupzR+KtZ9L0lgSz5+LKn4bI5JlcMkPT+rxrPVCZtfL4DwiOqdL6EzgIqIiIRq2uKWRdSdcgTcWnVOZyYnSba9ZXxF7kPmcS9D24nSxJtna1OjOKHxJAlZ4232ofDubV7zOjxPQ+PKJ57hKmhCnE4dl4QTjBSVbNLom32+1fFuV+VMlUARESmyxU16VBFji9K0ioWGGZx9Mt2NT9PdTZLfVKl3RtGzEeZhRLS4ZajkVx+eYmeVTDc5Sboc1iuTWctpUV1quvfsERHpbfHx0L2kZYubme0GnA7MjX/e3Y8oMF/SQrctwJ32xmnwxjTFGLeWcUefS3nO1p70g2PcPNEkCIUswB0bk5N08ePm8ZZ+tqqrlhf+Dm5hieLMvF57LOxImfyhenKShA83cs9Ng3QSfq/xCmm2P6rVZb+yHEDKk6xeXkLGuDXqTi0iIpKnkK6SFwHfAr4D9BebHRmp0lSuQp/yZ76B6sCYmY4bBnkdBlnsqLYuwN2mMMUZWjvrtvLVxY3fIiLSJiEVt83ufl7hOZGEuuOuIreJAKKfiSb7CAxjaSKvk07ttjSzGhYykUghk5OUNFxcOEW6pVa6/DLbHWdAsYYe/4RLMGScmKeVSrQZCmGek5Oknawna1fJutcIs1yP+0icQVVEJK2sva66VcgYt9+a2fvMbHsz27r8r/CcSUudmZzE60+ZnTa+FDdMg2s2hcadbe2rSnfQmniD46mZeS+XLo1N0skWb+tjG08nOMUh07VnWQ5gsNB08xi8ooTscWlpiMKzUlG7ZEYreUykUxNjk1fh4nkKfcgwAougiIh0SEiL21ujn6fHtjmwc/7ZkVDd8gQht+UAasaRJUk97VpiyVKp2WbJxxuV4sr3iytuAe7m4+biY+vC48y33HbJKVAo9+rjnLiVM8fxj/XkOalHcItS7IFMbYUx9XjWdMEG1UnY8og3RhVEERFpWXFz953akRFJo3vWA0i9AHeKcOEVtXxmmqt9nbjFbUi8ecxOErQpebQBkcQ/ErovWadrb5h+hniGq6TfUTt1bAHu2teJW8VLP9NOTtIOqriJiITr1R45IbNKjgXeC7wi2rQA+La7byowX9LFihynleREG5zJrnmYIhbgLks0Jm8YNRFVfx/dmfHhdDzTyms5gOLGuJW7AadYDqB2wYIMD2Ty2r+qrpJ5tCL26I2DiIh0RkhXyfOAscA3o9cnRdveWVSmpLW8ppVPakiSlfFf2ZYDSHKPFLqAQNY1PGpvSsuvHS9NgpBiDExeyzgMjkWzzBXT6nhLmi4H4IPdHsPHNFWPucqU18r3Sj4Hs9tZ7a+td7qodckaSTzBSM05kSpN4uNPvWpb0mLR7PoVsgB33V3I+diPhKIuIiLNhVTcXuLu+8ReX2dmdxaVIel+RSxmWDsBSFiY0s+WC3AHtsy1Cl9PohbCLm25aqabbxaH1fIKKWWp4MaPT1EzEoauqVhX2vFo9boJZ9y9evuRR+nKd4xbN5+NIiLdZSTPKtlvZi8ovzCzndF6bh3XLQWxk9kIrQhlzWOzyUnyiCuLeKtDrvG2iDI+MURwnOSb1y45BQpXVZlIPDlJe2T5WtM+0Kg3MU9Rk/WE5GXItvZnQ0REelxIi9vpwPVm9jClv0VzgbcXmisJ0pGuko2mGc84OUmSrlaV6fVbhKl92pJ2wgJqwsfjCZqavfxLnl0a605Okn/ErdIJTXHI5zJkNb5kQa+3QdQbYRg2OUl1973C13Fr8LpxuKHdC9NOdFN3W8IdrvdkNpcuyDke914v6yIi0lrIrJLXmtmuwO6U7iHuQ39DOq5but3ltwB3iskNAoNkPVaNwqfrHpb/cgB5K7UmNo84TetZmla65nnojnOgaGnWFqsXtkhZvov0Y9zi4zt9yLYkvPYpTeK81NmW98HXX10RkWBZ5znoVi27SprZAmB7d/+ru98JvBi4teiMSWtFjVtpnmb9m5R25iT4qX7GZoZGrQmJJz6oaSHMQ91Wh2JWGWiaTprJKErppM9sdfoj7242rFW6M92pg7+PgvOXR6kIXQNvBBZBERHpkJCukl8Efm9m5wKzgNcAbys0VzLipJlApHa2x9Zxp9N8cpKUkQ4TnXg4EKy3HqLVl9NqDIWV0xwjTvsQIM9s5N9IluPx6eZzUURE2iKkq+SVZvYe4GpgObCfuz/VKpyZXQC8FnjG3ffKnFOpkmQq9jyVnuQP3t1UKk9Zx7ilCNMqUObDU+l+VX7plc2WYnYSq4krS1eqeBeAXFvyPHCZhWgSiPAFl72yhEAljrR5jH4aOXzHw0C8m0dpXF/AcgBRyHbUbi2WTIIGt8p+pc1haSKSmuUAyuUyZcGoNxFMq1ZEp/65nPeR7/WHRCIi+cq/t1M3COkq+f+Ar1FagPtzwAIzOyYg7h8AR2XJnNR36aKlPLN6IxfdtoSDz7qOSxct7Vg+/vv39wFw3Df+mCkfv7j1cQA+9LNFwfFcd+8zAJxywa1Nw3zt2gcB+Oj/3ZEqjzc9uByA035xBwefdR1PPbeB5zb08f2bFgMEfweX//VJLkgYppXf/+1JvrXgIQCO+J8FucTpwLdvKMV5+Jfrx/n4s+tZu3Ez3//jI8H78tiz6znnmgcAOPqrf8ic10WPr+LRFes7eg4ULV5JK5/3Fy4MO+83bu7nU5fcBcBbL7ilkGPk0X//dfm9AJxw3p8Sp/PrO54A4H+v/nvQfpWPye//9hRv/0Gp1/6///YeAO5a+hx/emgFdy19LlG5qK0UXbpoKedeWyqrr/7KjYn36dJFS1m8fB1X3PVUbuVz0WOrKr/3cpkXEWmHSxct5eCzrmOnMy8fVtfUkK6SM4AD3H0DcLOZ/R74LnB5s0DufqOZzcueRYm7dNFSPnnJXfRHdxpLV23gk9HN2XH7zSo8/fL9TTkfGzaVVoZ4avXzqfJx/f2lytfajZsBWL6ur2U819z7NEAl7WVrN9YNc/ODKwB47vlNAKwIiDuuvK/fvuHhyralqzbwxKoNVW0erb6D8k3h535zN89vHgDgiefSHa/BvJUiPet39+UWJ8D9T63BHVY/v7lhnOv7NnP7o6sqxyCkDK7b2M+ilasq5TZteQFY9NhKAPqi/W73OdBuZsnP+8XL1rEm+g6h8TmS1f1PrcGBVRtK59jTa8LSKbdiXbpoKZ/+1V2V7Um+yy9ecW+l7K9cX0r/t3c+wYAnjyu+oH3tte3JFudVbSt1Ofzmgfyu0ZcuWsoli5ZUXvd6mRcRKVLtdX44XVNDukqeVvP60cAWNynAl6+8v1LQyjZs6ufLV97ftsJmOebjxzc/OmRbq3jKrVatwlx8+5Kgz7XS1z9Q9bpej6WQeMs3mVnyUnScNz+8Ysi22jhXb9hcqUCEpvvchk2JwzRy1d1Pt8xjrygfsqTn251LVg3ZVsQxuvmh1uWlEbPSfiUtw+VjUhsOqFTakubl8r+Wev/ncW0r4hr95SvvZ1N/PuePiEjZ7Y+t5KKFj7c93e22nMiHXrlLIevQltUuNxXXDffSaTWsuJnZTe5+SPT7j939pNjbfwb+IY8MmNmpwKkAc+bMySPKnvbEqg2Jtnd7Ppat3Zg4nmfWhIVZsa4vcdxZpIm3iLxkibPc8tksztoKWEi6/bV31AFhGim37uQR13Bglvx8W9/XX3d73sdoTUB5aaYd17OQuP7+9BoAZm4xPlWe4jcGRexTt1z3RaS3/Owvj3HJ7UuYMWV829Lc0NfPmo2bOemguWw9eVzb0o0bztfUZi1uk2O/71nzXm5VZHc/HzgfYP78+Rp+3cIO0yaytE7B2mHaxLak71G/oLzyMXPK+LqVt2bxbLPF+LqVt9ow0yePq1t5C81j0skAGsXbbDKJtN9bs7xlKQtTxo+pW3mLxznarG7lrVm6acI0Mm3i2LqVt3adA+1UPmJJz7dJ40bXrbzlfYxCyks95e6Faa4jSf9IhOyzGRy77w5MHj8mcZ5qi3UR1+hOX/dFpDcNDDg7TJvITZ84om1p/ujmR/jsr+9moIMzLg3na2qzyUmaHVFVsDrk9CN3Z+LY0VXbJo4dzelH7j4s83HSgXOHbGsVz9sP3ikozPH/MLS5O00ex4+uPk3qPbUIiXfCmOp48vje8o7zwJ23HrKtNs4tJo5hdE3fg1bpbpkiTCOv2mPblnnsJYYlPt/2mT1tyLYijtFBO0/PlM7pR+6eugzXhgMYVXNyhsYVH6eW5toWn/mziGv06UfuzthR+Zw/IiJlTvtnXRycXbvYdAZnoB66g91yL51Gs4rbNDN7vZm9Ifr9+OjfG4CprSI2s58DNwO7m9kSM3tHTnke0Y7bbxZfPH7vyk3wrGkT+eLxe7etT255WalyPqZNHAvAtluOT5WPV+w+Eyg9uYdSK1mreI540TYAlZNu5pT6aR8Y3VROjfIYEndceQKF9x72gsq2WdMm8qLtt2T2VhPZYkIpzztMndA03vLF6fPH7hkcpmXeop+fes2L2DKKc/uMcQLsuu0WjB5lTeOcOHY0+8/bqnKxDymDk8aNYf+5W1W+i+22TJ/XfXacBsD46Ma93edAJyQ97+dOn8TUiWMr3VAanSNZ7bbdFEYZbDUpug5sEZZO6WbBKvtVluS7/PQxL2Jm1L2nfB06dt9ZlQpdkrgcZ1R0bMt5KpfV7VuU1doW9XL4MaPyu0Yft9+sqvAjocyLSPHcB6997VIe19bJtSnL1+lZ0yZiDK9rarOukjcA/xT7/XWx925sFbG7vzlDvqSJ4/abxZevvJ+X7rw1Z79x347mY+Pmfj5x8V1c8r6DmZWhifnkg+byzQUPcc6b9uXlu84MCnP0XttxyaKlfO+U+by4TgtD2cdfvRv/79d38+UTX8wRLxzaWtPKobvP5JxrH+Ar/7wPr99vNm//wa2MGgUn7r8jX7nm7/zhE0cwuvZRfx2v22cHnl69kbOv/js3nnE4Y0a3XI2jpaP23o4Nm/r54u/u45qPHsrk8SETxTY3yuB9h+/CWb+7j2s/diiTxg2Nc+7Wk7j/qTUct+8OfP7YsGUa50yfxGv32Z7P/vpuLvvQIZn71B+w09YsXr6urV082i2+hlj5vD9w5+n87xv3aRl2wthR/Odxe/OuHy3kglNewt6zWz5vS5E/GGXGZ47Zg49ddCcXvucg5k6f3DpgzHH7zeIjF97Jaa/clY+8arewRIGj9tqef5i7FcecexOffM0L+cTFd7H3rKmseX4TT6x6nitOe3lwHgYGqGpKP26/WTy7ro9/v+wefv/hVzA1qpgm2adv3fAQc7aexPknz08UtpEX7ziNi24rTbj0xzN7t8yLSPuUH8a3Ux7rueah9oHYcNHwLs/d39bOjEgy1qnVhxsuwJ0uM+VwSR74BC/AXflcusvSYPDqRcbdPdHC1/HmeqvZllplcfDBfORRHEpdxup1LKj5jJUXgw6N12sWIM+Qx+inmXX8wt8O8XIWvAB3wPeYl/ha9MELcMe6JqadVaxqAe6q2cMs1blQe7QGz6sWC3C3685nJBR2EWmrAU9/DU6r9rpdlGazSg5nIQtwfyb2e/umnZERJc2JHBoma2Wh1TmfttI6HHTzrvXYtbiuvA5/UV1S8ow1+CFAAfkodRfKGEm9ePOPUkQkN+WHqu00KvChmNTXsOJmZmeY2UHACbHNNxefJQnRNU8QcspHmv0JDZP1WNULX3ran09cWcRbHfKLtHU+07TmmJHrAeiac6CNku5zO45RktbnfNMdun9pi9iAZztW9YLm/RRbtzgikreOTE4SpddglSBpodmAmPuBE4GdzewPwL3AdDPb3d3vb0vupKnO9JSs/3QmbctMba/HkGjKT2lahamdUShpHivN7DXha6NpFW295vqsLVl1Wx1yaB6rfQJWv3VjcGtw17gW6STi1d9/L6t7fAMOneO5lreG6dTEG95q5kO/wMBM1vvYkHwk3OHSda2mq2Rgtpr1lMzzuHdz67eIDE/u3sGuksVe1GrvFXtFs66SK4FPAQ8ChwHnRtvPNLM/FZwvaaE9o1dayy0XKS4cHW1xSxlv3hfInBuxBuNs8c2WxjUlSzjvMVft/mPTKfXGlIaHzTs39RLJfj1Key7VjkMzS5cTdxiVYa6gRtcIEZFu5nWeoRWtWyYnGa6atbgdBfwb8ALgbOBOYJ0mLekenRhbVZ6YIs/4ql+HTbwQEqZ2e9Kj1ahlbWj6reIZ+oGsfbvrtjpkirF+JPWObXxT6H4MiSaHyUlK8fb2lT/teK7aP8ZFHaUhLbQJWs2GNLiFptmiTKZR6rJTOzmJheWr6eQk+R35Xi/rItJ+5ZmB26l2wrfi02tPOu3S8Bmju3/K3V8JPAL8hFIlb6aZ3WRmv21T/mSESDVeLDBUj52zEtH3KmV5dBLu9j/uqraJSN4GvP3XvsFhLgV3lezRi2bIok9XuvutwK1m9l53P8TMZhSdMWkuyVTseSpPMz6Yj6xPTspdnOKvWuchJMzgtPHlcAnHvdQsVVB+7dHG0EUkCxnjVsmbxbZlixMGByo3O7aVzyRIszzmKo+lC+LHs0evy4NqZvwyC2yVptxtsOCxBF7dXTdJKtVdQJOPl6zq0hs7V0vHKEFGKD91rs1f+b0W5zeO1XkGmiYfIiLt1IlLVLlbuq6P6bTs1e/uZ8RenhJtW15UhmRkSjMyJXiMW9bxN2rb6VL6XqQk69//AR86OUm30U2OiOStNPylM5OTDLTtotbd1/akEg3Hdvc7i8qIJJPkCXWeamery9rknWaBRK965t467qznbG1rQtp1T/KcNr2q1SHHFbg94AbWo/FAyRbgrl66INMC3LHWlZFwM1t7viVplW7H36s0Lcle0z0nzY1DvYW/y61wSa9H5Vbkqvhj7zUN22Dcb973QiOgqItI2xWzhmUzOd6yNNWr18wM82iJ5Kdy3UhwpoXOTBR6A9YqnUZ6ufLQzfvWW8/Q6svr8Bc3OUmecaWc6IbsXUGLGqDfxaePiEjmNSzTaPfkJL1GFbdhqlumQs8rG6mmAw+OO/+ukvGn/cHx5DBtev048xUy1X/p7aQp57tQc5ecAoWrHuMWvtNFlI266WDpWswa/J4o3SEb05Wxul12Eh7rIdtyPvqaVVJE8hbSyyZvlYfpbbqm9dq9gipuw1inJidJsr1lfNHP0Ik+4mkNnozNpycZvEgkzFttOmn3sYDpy+svB5DD3HpB0/YPbuzEAtyDvQCTd4kbbkIWm64bbsg0/TllaEheggrM0HB14wpMM2Bb4v1t8tQ5ZAHuhu+psiUiXcwZOjFT0drWVbJHr7+quA1T3fIAoZNPMoKXA8g6xq3BtiInVAmOL+dWrFKcrctXfCbB4HhzbgHqtadojdTOvpg2bFHy+F5TZbPRGLcUcZVuXtLvRaNWeRGRbjZQb4BvwcrX2h6tVxVOFbdhrCMLcFP/ZjBtTpIuZh0SR+PtSScs8LqvEy/AHbupbBR3UllaLJLEW3fxcK/+RFC8OXzP9cL2+oW//uLtQQFr/hYXc6DSfq/1ZjILn+im9TFJurcDdSYcCp14qdl1ONcxgD1e1kWk/dJOtpZFOb12zSrZa8/QVHGTrpBqjFtgmMwtbr121vcIfS9SlsfkJCpPIjISdayrpB5GpVJoxc3MjjKz+83sQTM7s8i0RpwEU7HnqbQAdzwb5SbvdLmpTO2eYJr42has5iPc0k9BP/j56vClJREs1k+7xRP5ciw22KEqrwW4sdbHIVm8VO9b3bF0g93jwvfDUy/UXC99KI9x62215xsWdq451ZN3FLj+dqrv1fEh15Gkeaw3QU55AqCk1yPHh3SVDB3b2qinUbLzo7VeH88pIu3XkXXcEsxpIEMVVnEzs9HAN4CjgT2AN5vZHkWlJ8Nbuha3No1x05P47qTvRSJZK0gD5RpoF9PTaRHJW71u4kUr+oHikPR67CauyBa3A4AH3f1hd+8DfgEcW2B6I0oni2H1wrk5xZlPNIVoPDlJ8nhyn0gkYOr+xHESMjlJ8klR8l4OoZvLTJ7qLXgfGq4df69KLckZJ+pJ8+CmNg6i8pVmn31o2cxcVnvsZkFEek8nuolrHbdsxhQY9yzg8djrJcBLC0xvRDEzrrnnaQ78r2vbmu6z6/vYbdspQ7a/8ds3M2ZU8ucAff0DwOAsQx//5Z1MGDO6aZhNNWHe/ePbGDd6aNobNvVXfe7MS+7i87+9Jzhv6zZurgr/X1fcyznXPMCz6/rYZ8eplYvdYV9e0HRGurVRPDB4M3j4/zQPE5q3+M3rkefcyOiMV+DVz29i9KjBdbleXSfOlev7KpWwS+9YyoL7l7WMd8W6vqqb6mO//kfGpOxYH/9elq/d2PZzoJ2Wr93I9lMnVl6bGVcHnPerNvSx3ZYTKuXt1AbnSFarn9/E6FjX2n/5zp+DrgPPbdjE5PGDf34M+PHNj3DpoqUtw1bKfqzr8bnXPhBtK8X1yIr1icpFX//AkHEe5X065ms3NT2vVqzbyEEvmDFkuwE3P7Qit/K5LnYd6eUyLyLt8+y6PvadM62taZb/FJ3y/VsYW8DfpbLyPWCvPUIrsuJW71gNqV+b2anAqQBz5swpMDu95YNH7MKfHlzRkbRfsdvMyu8H7zKDkw+ay8ZNA6njmzR+NCfOn80za55n9YbNrQMAE8eN5p0v3wnHWb+xv+Hnpk0ay6v22JZ3v2JnVq3flDhvc6ZPYu70SXzwiF14ZvXGyvZ/3GNbdtt2Co+sWE9/f+vHRjvPnMyoUcaRe27H4hXrgsK0MmuriWwxYSyvfNG23P/UGjbnECfAXrO25PAXbsPfn64fpxmcuP+O7LnDltz+6KqgOM3gxPmz2XbLCZx04Fz6NqcvL1Da90N3m8mkcaN7/qnd6/bZofJ7kvP+gJ22Zt8503jbwfOaniNZ7TlrSw7ceTpvPWguzye4DsRvFs48+oXc9+Sa4LCzt5rIlPFjmDh9Mu859AWsXNfH2DHGYbvPZOcZk5k8LtmftlGj4Nh9Z1VtO2z3mbzlwDls2ty6gB2993ZDtr3n0J25/r7WDzWS2NQ/wPixoxkY6PFCLyJt86o9tm1revvP3ZpTXjaPDX3F/V0q22ryOHbcelLh6bSTFTWlvJkdBHzO3Y+MXn8SwN2/2CjM/PnzfeHChYXkR0REREREpNuZ2W3uPr92e5Fj3G4FdjWzncxsHPAm4DcFpiciIiIiItKTCusq6e6bzewDwJXAaOACd7+7qPRERERERER6VZFj3HD3K4ArikxDRERERESk1xU2xi0NM1sGPNrpfNQxA1je6UyI5EhlWnqNyrT0GpVp6TUq0+HmuvvM2o1dVXHrVma2sN4AQZHhSmVaeo3KtPQalWnpNSrT2RU5OYmIiIiIiIjkQBU3ERERERGRLqeKW5jzO50BkZypTEuvUZmWXqMyLb1GZTojjXETERERERHpcmpxExERERER6XKquDVhZkeZ2f1m9qCZndnp/IjEmdkFZvaMmf0ttm1rM7vazB6Ifm4Ve++TUVm+38yOjG3f38zuit4718ws2j7ezP4v2v4XM5vX1h2UEcfMdjSz683sXjO728xOi7arXMuwZGYTzOwWM7szKtOfj7arTMuwZmajzWyRmV0WvVaZbgNV3Bows9HAN4CjgT2AN5vZHp3NlUiVHwBH1Ww7E7jW3XcFro1eE5XdNwF7RmG+GZVxgPOAU4Fdo3/lON8BrHT3XYCvAF8qbE9ESjYDH3P3FwEHAu+Pyq7KtQxXG4Ej3H0fYF/gKDM7EJVpGf5OA+6NvVaZbgNV3Bo7AHjQ3R929z7gF8CxHc6TSIW73wg8W7P5WOCH0e8/BI6Lbf+Fu29098XAg8ABZrY9sKW73+ylAa8/qglTjuuXwCvLT8NEiuDuT7r77dHvayjdFMxC5VqGKS9ZG70cG/1zVKZlGDOz2cAxwHdjm1Wm20AVt8ZmAY/HXi+Jtol0s23d/Uko3QQD20TbG5XnWdHvtdurwrj7ZuA5YHphOReJibrG7Af8BZVrGcaiLmV3AM8AV7u7yrQMd+cAZwADsW0q022giltj9Wr2moJThqtG5blZOdc5IB1hZlOAi4EPu/vqZh+ts03lWrqKu/e7+77AbEotDXs1+bjKtHQ1M3st8Iy73xYapM42lemUVHFrbAmwY+z1bOCJDuVFJNTTUfcDop/PRNsblecl0e+126vCmNkYYCpDu2aK5MrMxlKqtP3U3S+JNqtcy7Dn7quABZTG8ahMy3B1MPBPZvYIpWFER5jZT1CZbgtV3Bq7FdjVzHYys3GUBlb+psN5EmnlN8Bbo9/fCvw6tv1N0UxNO1EaBHxL1J1hjZkdGPUfP7kmTDmuE4DrXAs/SoGiMvg94F53Pzv2lsq1DEtmNtPMpkW/TwT+EbgPlWkZptz9k+4+293nUbo3vs7d34LKdFuM6XQGupW7bzazDwBXAqOBC9z97g5nS6TCzH4OHAbMMLMlwL8BZwEXmtk7gMeAEwHc/W4zuxC4h9LMfe939/4oqvdSmqFyIvC76B+UbqB/bGYPUnrS9aY27JaMbAcDJwF3RWOCAD6FyrUMX9sDP4xm0RsFXOjul5nZzahMS2/RdboNTBVYERERERGR7qaukiIiIiIiIl1OFTcREREREZEup4qbiIiIiIhIl1PFTUREREREpMup4iYiIiIiItLlVHETEZGuY2bbmtnPzOxhM7vNzG42s9cXnObnzOzjBcT7YTObFHu9Nu80RESk96niJiIiXSVajPVS4EZ339nd96e0js/sjmYsvQ8Dk1p9SEREpBlV3EREpNscAfS5+7fKG9z9UXf/GoCZjTazL5vZrWb2VzN7d7T9MDNbYGa/NLP7zOynUSUQM9vfzG6IWu+uNLPtm2XAzF5gZr+PPv8HM3thtP0HZnaumf0pag08Ido+ysy+aWZ3m9llZnaFmZ1gZh8CdgCuN7PrY/F/wczuNLM/m9m2OR8/ERHpQaq4iYhIt9kTuL3J++8AnnP3lwAvAd5lZjtF7+1HqYVrD2Bn4GAzGwt8DTghar27APhCizycD3ww+vzHgW/G3tseOAR4LXBWtO14YB6wN/BO4CAAdz8XeAI43N0Pjz47Gfizu+8D3Ai8q0VeREREGNPpDIiIiDRjZt+gVFHqiyprrwZeXG7tAqYCuwJ9wC3uviQKdwelytQqYC/g6qgBbjTwZJP0pgAvAy6KPg8wPvaRS919ALgn1lp2CHBRtP2peOtaHX3AZdHvtwGvavJZERERQBU3ERHpPncDbyi/cPf3m9kMYGG0ySi1hl0ZD2RmhwEbY5v6Kf2dM+Budz8oMP1RwCp337fB+/E0rOZniE3u7jV5FBERaUpdJUVEpNtcB0wws/fGtsUn97gSeG/UBRIz283MJjeJ735gppkdFH1+rJnt2ejD7r4aWGxmJ0afNzPbp0WebwLeEI112xY4LPbeGmCLFuFFRESaUsVNRES6StQadRxwqJktNrNbgB8Cn4g+8l3gHuB2M/sb8G2atFq5ex9wAvAlM7sTuINSV8hm/hV4R/T5u4FjW3z+YmAJUM7PX4DnovfOB37XovukiIhIUzbYW0NERETSMrMp7r7WzKYDtwAHu/tTnc6XiIj0BvWrFxERycdlZjYNGAf8hyptIiKSJ7W4iYiIiIiIdDmNcRMREREREelyqriJiIiIiIh0OVXcREREREREupwqbiIiIiIiIl1OFTcREREREZEup4qbiIiIiIhIl/v/5DmoY3gdyqMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1080x144 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\" Visualization\"\"\"\n",
    "exonStarts = refseq_df.loc[refseq_df['gene_id'] == 'PDGFRB']['start']\n",
    "exonEnds = refseq_df.loc[refseq_df['gene_id'] == 'PDGFRB']['end']\n",
    "# exonStarts = refseq_df.loc[refseq_df['gene_id'] == 'TRIM17']['start']\n",
    "# exonEnds = refseq_df.loc[refseq_df['gene_id'] == 'TRIM17']['end']\n",
    "\n",
    "cutoff = 0.5\n",
    "\n",
    "geneStart = np.min(exonStarts)\n",
    "geneEnd = np.max(exonEnds)\n",
    "geneLen = geneEnd - geneStart + 1\n",
    "collapse = np.zeros(geneLen)\n",
    "\n",
    "for i, (start, end) in enumerate(zip(exonStarts, exonEnds)):\n",
    "    collapse[(start - geneStart) : (end - geneStart + 1)] += 1\n",
    "\n",
    "n_iso = collapse.max()\n",
    "overlaps = np.where(collapse >= n_iso * cutoff)[0]\n",
    "\n",
    "noDirectionBounds = np.where((overlaps - np.append(overlaps[1:], overlaps[-1]+3)) < -1)[0]\n",
    "cExonStarts, cExonEnds = [], []\n",
    "for i, bound in enumerate(noDirectionBounds):\n",
    "    if i == 0:\n",
    "        cExonStarts.append(geneStart + overlaps[0])\n",
    "    else:\n",
    "        cExonStarts.append(geneStart + overlaps[noDirectionBounds[i-1] + 1])\n",
    "\n",
    "    cExonEnds.append(geneStart + overlaps[bound])\n",
    "\n",
    "print(cExonStarts)\n",
    "print(cExonEnds)\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "plt.figure(figsize = (15, 2))\n",
    "plt.plot(collapse)\n",
    "for x1, x2 in zip(cExonStarts, cExonEnds):\n",
    "    plt.plot([x1-geneStart , x2-geneStart], [n_iso, n_iso], 'r')\n",
    "plt.scatter(overlaps[noDirectionBounds], len(noDirectionBounds) * [1])\n",
    "plt.title('Finding Constitutive Exons')\n",
    "plt.xlabel('Gene length')\n",
    "plt.ylabel('#Exon Overlap')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating the jobfiles within the directory for each chromosome, making one .sh file to run ppDesigner and run that file from the command line\n",
    "Parameters that are set manually: max overlap of 20 bases between probes. Large exon size of 150. Max target region 40. \n",
    "\n",
    "What needs to be run from the command line:\n",
    "\n",
    "(base) kian@genome-voyager:~/Codes/DART-FISH/ProbeProduction/221010_WholeGenome_overlappingProbes$ ./ppDesignerCommands.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "from time import time \n",
    "import os\n",
    "import stat\n",
    "\n",
    "\n",
    "def getJobFile(targetFile, ppd_dir):   \n",
    "    jobFile = \"\"\"#!/usr/bin/perl -w\n",
    "#Use probe script with this file as ARGV[0].\n",
    "\n",
    "our $softwareDir='{1}';\n",
    "our $HsDir='/media/Scratch_SSD_Voyager/Kian/genomes/hg38/byChr/';\n",
    "\n",
    "our $exon_info_file='{0}';\n",
    "\n",
    "our $primerMaxLen=22;\n",
    "our $primerMinLen=17;\n",
    "our $H1_plus_H2_Len = 39; # sum of arm lengths-1\n",
    "\n",
    "our $primerMaxTm = 65;\n",
    "our $primerMinTm = 55;\n",
    "\n",
    "our $flankingLen = 0; # 221006: from 25 to 0;\n",
    "\n",
    "our $largeExonSize = 150;\t\t# 210128 Kian: changed from 500 to 150\n",
    "\n",
    "our $targetMinLen=1;\n",
    "our $targetMaxLen=1;\n",
    "\n",
    "our $arewebisulfite=0;\n",
    "our $using_unafold=1;\n",
    "\n",
    "our $jumpFromPrevious=0; # for targets>$largeExonSize, if a probe is found in a chunk, where should should the next chunk start from. 0 corresponds to the ligation junction of the previous probe\n",
    "our $jumpIf0Found=$primerMaxLen; # if a probe is not found in a chunk, how much to move the next chunk forward.\n",
    "our $maxOverlap=20; # maximum number of bases that can overlap between a probe pair\n",
    "\n",
    "\n",
    "our $NNParamFile = \"$softwareDir/NN_param.txt\"; # path to NN_param.txt\n",
    "\n",
    "\n",
    "#DO NOT CHANGE THE BELOW LINES.\n",
    "eval `cat $softwareDir/probe_parameters.pl` or die 'couldnt parse file';\n",
    "eval `cat $softwareDir/load_inputs.pl` or die 'couldnt parse file';\n",
    "eval `cat $softwareDir/get_sequence.pl` or die 'couldnt parse file';\n",
    "eval `cat $softwareDir/get_probes.pl` or die 'couldnt parse file';\n",
    "eval `cat $softwareDir/neural_net.pl` or die 'couldnt parse file';\n",
    "eval `cat $softwareDir/output_text.pl` or die 'couldnt parse file';\n",
    "\"\"\".format(targetFile, ppd_dir)\n",
    "    return jobFile\n",
    "\n",
    "def writeJobFile(jfile, addr):\n",
    "    with open(addr, 'w') as writer:\n",
    "        writer.writelines(jfile)\n",
    "\n",
    "def run_ppDesigner(command, outfile):\n",
    "    print(outfile)\n",
    "    with open(outfile, 'w') as outer: \n",
    "        ppd_proc = sp.Popen(command, stdout = sp.PIPE)\n",
    "        commandOut = ppd_proc.communicate()\n",
    "    return commandOut\n",
    "\n",
    "def makeExecutable(file):\n",
    "    st = os.stat(file)\n",
    "    os.chmod(file, st.st_mode | stat.S_IEXEC)\n",
    "       \n",
    "ppDesignerPath = os.path.join(os.getcwd(), 'opt/ppDesigner/src/ppDesigner.pl')\n",
    "for chrm in chrms:\n",
    "    jfile = getJobFile(os.path.join(savingdir, chrm, '{}.target'.format(chrm)), os.path.dirname(ppDesignerPath))\n",
    "    writeJobFile(jfile, os.path.join(savingdir, chrm, '{}.jfile'.format(chrm)))\n",
    "\n",
    "\n",
    "# with open(os.path.join(savingdir, 'ppDesignerCommands.sh'), 'w') as ppdc:\n",
    "#     ppdc.writelines(\"\"\"#!/bin/bash\n",
    "\n",
    "# for indx in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y\n",
    "# do\n",
    "#     cd chr$indx\n",
    "#     {1} chr${{indx}}_largeExon{0}.jfile > chr${{indx}}_largeExon{0}.ppd &\n",
    "#     cd ..\n",
    "# done\n",
    "\n",
    "# \"\"\".format(largeExon, ppDesignerPath))    \n",
    "\n",
    "with open(os.path.join(savingdir, 'ppDesignerCommands.sh'), 'w') as ppdc:\n",
    "    ppdc.writelines(\"\"\"#!/bin/bash\n",
    "\n",
    "chrlist='1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M'\n",
    "runPPD () {{\n",
    "    local indx=$1\n",
    "    cd chr$indx\n",
    "    {0} chr${{indx}}.jfile > chr${{indx}}.ppd\n",
    "    cd ..\n",
    "}}\n",
    "    \n",
    "for indx in $chrlist; do runPPD \"$indx\" & done\n",
    "\n",
    "\"\"\".format(ppDesignerPath))    \n",
    "\n",
    "\n",
    "    \n",
    "makeExecutable('ppDesignerCommands.sh')\n",
    "makeExecutable(ppDesignerPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making zero-gap probes\n",
    "Since the current version of ppDesigner doesn't produce 0-gap probes, Matt had a file (ConvertToZeroGapProbe.pl) to read ppDesigner's output and add a base to the arm with lower Tm. Here, I make that file again and change it write an executable to run it for each chromosome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ConvertToZeroGapProbe = \"\"\"#!/usr/bin/perl\n",
    " use strict;\n",
    " use warnings;\n",
    " \n",
    "#print \"Go from 1 gap to 0 gap\\\\nAdds target base to arm with lowest Tm\\\\nRemove any probes targeting softmasked regions (lowercase)\";\n",
    "my $onegap_filename = $ARGV[0];\n",
    "open(DAT, $onegap_filename) || die(\"Error opening 1gap probe file\");\n",
    "\n",
    "while(<DAT>)\n",
    "{\n",
    "\tmy @line = split('\\\\t',$_);\n",
    "\tif ($#line > 3)\n",
    "\t{\n",
    "        my $donorLoc = 0;\n",
    "        my $midbaseStart = $line[3];\n",
    "\t\tmy $leftArm = $line[5];\n",
    "\t\tmy $leftTm = $line[6];\n",
    "\t\tmy $rightArm = $line[7];\n",
    "\t\tmy $rightTm = $line[8];\n",
    "\t\tmy $annealRegion = $line[11];\n",
    "\t\t\n",
    "\t\tif ($leftTm <= $rightTm) #if leftTm is less than or equal to rightTm\n",
    "\t\t{\n",
    "\t\t\t$leftArm = $leftArm . substr($annealRegion,length($leftArm),1); #add base to left arm\n",
    "            $donorLoc = $midbaseStart + 1;\n",
    "\t\t}\n",
    "\t\telse #else if rightTm is greater than leftTm\n",
    "\t\t{\n",
    "\t\t\t$rightArm = substr($annealRegion,length($leftArm),1) . $rightArm; #add base to right arm\n",
    "            $donorLoc = $midbaseStart;\n",
    "\t\t}\n",
    "\t\t\t\n",
    "\t\tmy $newLine = $line[0];\n",
    "\t\tfor (my $n = 1;$n<$#line;$n++) \n",
    "\t\t{\n",
    "            if ($n == 3){\n",
    "                $newLine = $newLine.\"\\\\t\".$donorLoc;\n",
    "            }\n",
    "            elsif ($n == 4){\n",
    "                $newLine = $newLine.\"\\\\t\".$donorLoc;\n",
    "            }\n",
    "\t\t\telsif ($n == 5)\n",
    "\t\t\t{\n",
    "\t\t\t\t$newLine = $newLine.\"\\\\t\".$leftArm;\n",
    "\t\t\t}\n",
    "\t\t\telsif ($n == 7)\n",
    "\t\t\t{\n",
    "\t\t\t\t$newLine = $newLine .\"\\\\t\".$rightArm;\n",
    "\t\t\t}\n",
    "\t\t\telse\n",
    "\t\t\t{\n",
    "\t\t\t\t$newLine = $newLine.\"\\\\t\".$line[$n];\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t\t\n",
    "\t\tif($line[11] eq uc $line[11]) #if all uppercase\n",
    "\t\t{\n",
    "\t\t\tprint $newLine.\"\\\\n\";\t\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "}\n",
    "close(DAT);\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "with open(os.path.join(savingdir, 'ConvertToZeroGapProbe.pl'), 'w') as writer:\n",
    "    writer.writelines(ConvertToZeroGapProbe)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write an executable for converting whole genome probes to 0-gap\n",
    "\n",
    "Run this in the command line:\n",
    "(base) kian@genome-voyager:~/Codes/DART-FISH/ProbeProduction/221010_WholeGenome_overlappingProbes$ ./execute0gapConverter.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import stat\n",
    "\n",
    "def makeExecutable(file):\n",
    "    st = os.stat(file)\n",
    "    os.chmod(file, st.st_mode | stat.S_IEXEC)\n",
    " \n",
    "with open(os.path.join(savingdir, 'execute0gapConverter.sh'), 'w') as ppdc:\n",
    "    ppdc.writelines(\"\"\"#!/bin/bash\n",
    "\n",
    "for indx in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M\n",
    "do\n",
    "perl {0}/ConvertToZeroGapProbe.pl {0}/chr${{indx}}/chr${{indx}}.ppd > {0}/chr${{indx}}/chr${{indx}}_0gap.ppd &\n",
    "wait\n",
    "done\n",
    "\"\"\".format(savingdir))\n",
    "makeExecutable(os.path.join(savingdir, 'execute0gapConverter.sh'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making Fasta files for alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrm in chrms:#[0:1]:\n",
    "    ppd_0g = pd.read_csv(os.path.join(savingdir, chrm, '{0}_0gap.ppd'.format(chrm)), sep = '\\t', header = None)\n",
    "    with open(os.path.join(savingdir, chrm, 'allProbes_{0}.fa'.format(chrm)), 'w') as writer:\n",
    "        for i, row in ppd_0g.iterrows():\n",
    "            writer.write('>{0}_mid{1}\\n'.format(row[0], row[3])) # write the name\n",
    "            writer.write(\"{}\\n\".format(row[11]))\n",
    "#             if i > 20:\n",
    "#                 break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ND1_chrM_1</td>\n",
       "      <td>chrM:000003307-4262</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>ACCCATGGCCAACCTCCT</td>\n",
       "      <td>64</td>\n",
       "      <td>ACTCCTCATTGTACCCATTCTA</td>\n",
       "      <td>57</td>\n",
       "      <td>2.560457</td>\n",
       "      <td>W</td>\n",
       "      <td>ACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTA</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ND1_chrM_1</td>\n",
       "      <td>chrM:000003307-4262</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>ATCGCAATGGCATTCCTA</td>\n",
       "      <td>57</td>\n",
       "      <td>ATGCTTACCGAACGAAAAATTC</td>\n",
       "      <td>58</td>\n",
       "      <td>2.533262</td>\n",
       "      <td>W</td>\n",
       "      <td>ATCGCAATGGCATTCCTAATGCTTACCGAACGAAAAATTC</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ND1_chrM_1</td>\n",
       "      <td>chrM:000003307-4262</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>CGAACGAAAAATTCTAGGCT</td>\n",
       "      <td>57</td>\n",
       "      <td>ATATACAACTACGCAAAGGC</td>\n",
       "      <td>55</td>\n",
       "      <td>2.523006</td>\n",
       "      <td>W</td>\n",
       "      <td>CGAACGAAAAATTCTAGGCTATATACAACTACGCAAAGGC</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ND1_chrM_1</td>\n",
       "      <td>chrM:000003307-4262</td>\n",
       "      <td>0</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>CCCAACGTTGTAGGCCCCT</td>\n",
       "      <td>63</td>\n",
       "      <td>ACGGGCTACTACAACCCTTCG</td>\n",
       "      <td>64</td>\n",
       "      <td>2.550944</td>\n",
       "      <td>W</td>\n",
       "      <td>CCCAACGTTGTAGGCCCCTACGGGCTACTACAACCCTTCG</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ND1_chrM_1</td>\n",
       "      <td>chrM:000003307-4262</td>\n",
       "      <td>0</td>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>ACCCTTCGCTGACGCCA</td>\n",
       "      <td>64</td>\n",
       "      <td>TAAAACTCTTCACCAAAGAGCCC</td>\n",
       "      <td>62</td>\n",
       "      <td>2.630296</td>\n",
       "      <td>W</td>\n",
       "      <td>ACCCTTCGCTGACGCCATAAAACTCTTCACCAAAGAGCCC</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                    1   2    3    4                     5   6   \\\n",
       "0  ND1_chrM_1  chrM:000003307-4262   0   19   19    ACCCATGGCCAACCTCCT  64   \n",
       "1  ND1_chrM_1  chrM:000003307-4262   0   59   59    ATCGCAATGGCATTCCTA  57   \n",
       "2  ND1_chrM_1  chrM:000003307-4262   0   87   87  CGAACGAAAAATTCTAGGCT  57   \n",
       "3  ND1_chrM_1  chrM:000003307-4262   0  126  126   CCCAACGTTGTAGGCCCCT  63   \n",
       "4  ND1_chrM_1  chrM:000003307-4262   0  156  156     ACCCTTCGCTGACGCCA  64   \n",
       "\n",
       "                        7   8         9  10  \\\n",
       "0   ACTCCTCATTGTACCCATTCTA  57  2.560457  W   \n",
       "1   ATGCTTACCGAACGAAAAATTC  58  2.533262  W   \n",
       "2     ATATACAACTACGCAAAGGC  55  2.523006  W   \n",
       "3    ACGGGCTACTACAACCCTTCG  64  2.550944  W   \n",
       "4  TAAAACTCTTCACCAAAGAGCCC  62  2.630296  W   \n",
       "\n",
       "                                         11   12  \n",
       "0  ACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTA  0.2  \n",
       "1  ATCGCAATGGCATTCCTAATGCTTACCGAACGAAAAATTC  0.2  \n",
       "2  CGAACGAAAAATTCTAGGCTATATACAACTACGCAAAGGC  0.2  \n",
       "3  CCCAACGTTGTAGGCCCCTACGGGCTACTACAACCCTTCG  0.2  \n",
       "4  ACCCTTCGCTGACGCCATAAAACTCTTCACCAAAGAGCCC  0.2  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppd_0g.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running BWA-MEM\n",
    "(base) kian@genome-voyager:~/Codes/DART-FISH/ProbeProduction/221010_WholeGenome_overlappingProbes$ ./runBWA.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "runAligner = \"\"\"#!/bin/bash\n",
    "\n",
    "chrlist='1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M'\n",
    "runBWA () {{\n",
    "    local indx=$1\n",
    "    cd chr$indx\n",
    "    ~/packages/bwa_201123/bwa/bwa mem -k 8 -T 0 /media/Scratch_SSD_Voyager/Kian/genomes/hg38/NoAlt/hg38_noAlt.fa allProbes_chr${{indx}}.fa > bwamem_chr${{indx}}.sam \n",
    "    cd ..\n",
    "}}\n",
    "    \n",
    "for indx in $chrlist; do runBWA \"$indx\" & done\n",
    "wait\n",
    "\"\"\".format()\n",
    "with open(os.path.join(savingdir, 'runBWA.sh'), 'w') as writer:\n",
    "    writer.writelines(runAligner)\n",
    "makeExecutable(os.path.join(savingdir, 'runBWA.sh'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing non-unique alignments\n",
    "Remove probes that have alignments with mapping quality (MAPQ) <= 30, or with secondary alignment (existence of XA tag), or with not-matching aligned chromosome.\n",
    "\n",
    "Since the reference has all the alternate contigs removed, we don't have any worries about spurious entries with MAPQ==0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from copy import deepcopy\n",
    "# import re\n",
    "\n",
    "# def getXAchr(xatag):\n",
    "#     if pd.isna(xatag):\n",
    "#         return []\n",
    "#     pattern = r\"([a-zA-Z0-9_-]+),([0-9-]+),([a-zA-Z0-9_-]+),([a-zA-Z0-9_-]+);\"\n",
    "#     matches = re.findall(pattern, xatag)\n",
    "#     return [match[0] for match in matches]\n",
    "\n",
    "# def handleNoAltSam(df, chrm):\n",
    "#     \"\"\" A function to handle aligned probes whose gene isn't located on alt contigs\"\"\"\n",
    "#     sam_out = deepcopy(df)\n",
    "    \n",
    "#     # Removing probes with non-unique alignment\n",
    "#     badprobes1 = sam_out.loc[sam_out['MAPQ'].astype(int) <= 20 ] \n",
    "#     sam_out = sam_out.loc[sam_out['MAPQ'].astype(int) > 20] # removing probes with low mapping quality \n",
    "#     badprobes2 = sam_out.loc[~sam_out['XA'].isna()]\n",
    "#     sam_out = sam_out.loc[sam_out['XA'].isna()] # removing probes with secondary alignment\n",
    "#     badprobes3 = sam_out.loc[~sam_out['RNAME'].str.startswith(chrm)]\n",
    "#     sam_out = sam_out.loc[sam_out['RNAME'].str.startswith(chrm)] # removing probe whose chromosome doesn't match\n",
    "#     return sam_out, pd.concat([badprobes1, badprobes2, badprobes3], ignore_index = True)\n",
    "\n",
    "# def handleAltSam(df, chrm):\n",
    "#     \"\"\" A function to handle aligned probes whose gene IS located on alt contigs.\n",
    "#         These probes will likely have secondary alignments to their alt locations\"\"\"\n",
    "#     sam_out = deepcopy(df)\n",
    "#     badprobes1 = sam_out.loc[~sam_out['RNAME'].str.startswith(chrm)]\n",
    "#     sam_out = sam_out.loc[sam_out['RNAME'].str.startswith(chrm)] # removing probe whose chromosome doesn't match\n",
    "#     XAchrms = sam_out['XA'].apply(getXAchr)\n",
    "#     allAlignedChrms = XAchrms + sam_out['chr'].apply(lambda x : [x]) # the chromosome of the primary and all secondary alignments for each gene/probe\n",
    "#     allAltChrms = sam_out['gene'].apply(lambda x: genesWithAlts[x]) # the all contigs (ref and alt) of each gene/probe\n",
    "\n",
    "#     df_temp = pd.concat([allAlignedChrms, allAltChrms], axis = 1)\n",
    "#     alignedIsAlt = df_temp.apply(lambda x : all([algn in x.iloc[1] for algn in x.iloc[0]]), \n",
    "#                                  axis = 1) # check to see if the aligned chrms are all in the contigs for each gene/probe\n",
    "#     badprobes2 = sam_out.loc[~alignedIsAlt]\n",
    "#     sam_out = sam_out.loc[alignedIsAlt]\n",
    "#     return sam_out, pd.concat([badprobes1, badprobes2], ignore_index = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/media/Home_Raid1_Voyager/kian/.local/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3331: DtypeWarning: Columns (1,3,4) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1: Removed 21824 probes out of 199536 (10.94%)\n",
      "chr2: Removed 10535 probes out of 128387 (8.21%)\n",
      "chr3: Removed 9236 probes out of 111770 (8.26%)\n",
      "chr4: Removed 4789 probes out of 67072 (7.14%)\n",
      "chr5: Removed 7906 probes out of 88687 (8.91%)\n",
      "chr6: Removed 68610 probes out of 216937 (31.63%)\n",
      "chr7: Removed 12556 probes out of 94768 (13.25%)\n",
      "chr8: Removed 6776 probes out of 67852 (9.99%)\n",
      "chr9: Removed 8597 probes out of 77921 (11.03%)\n",
      "chr10: Removed 7289 probes out of 77091 (9.46%)\n",
      "chr11: Removed 9778 probes out of 108807 (8.99%)\n",
      "chr12: Removed 7737 probes out of 98451 (7.86%)\n",
      "chr13: Removed 2602 probes out of 35295 (7.37%)\n",
      "chr14: Removed 6571 probes out of 65082 (10.10%)\n",
      "chr15: Removed 15439 probes out of 84819 (18.20%)\n",
      "chr16: Removed 10548 probes out of 82422 (12.80%)\n",
      "chr17: Removed 19933 probes out of 127692 (15.61%)\n",
      "chr18: Removed 2212 probes out of 31238 (7.08%)\n",
      "chr19: Removed 31460 probes out of 144211 (21.82%)\n",
      "chr20: Removed 2509 probes out of 47439 (5.29%)\n",
      "chr21: Removed 3414 probes out of 22935 (14.89%)\n",
      "chr22: Removed 7583 probes out of 51573 (14.70%)\n",
      "chrX: Removed 9620 probes out of 72900 (13.20%)\n",
      "chrY: Removed 3728 probes out of 8699 (42.86%)\n",
      "chrM: Removed 178 probes out of 483 (36.85%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QNAME</th>\n",
       "      <th>FLAG</th>\n",
       "      <th>RNAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>MAPQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>RNEXT</th>\n",
       "      <th>PNEXT</th>\n",
       "      <th>TLEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>NM</th>\n",
       "      <th>MD</th>\n",
       "      <th>AS</th>\n",
       "      <th>XS</th>\n",
       "      <th>XA</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>WASH7P_chr1_3_mid114</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15891</td>\n",
       "      <td>60</td>\n",
       "      <td>39M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCTAGCAGAGTGGCCAGCCACCGGAGGGGTCAACCACTT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:39</td>\n",
       "      <td>AS:i:39</td>\n",
       "      <td>XS:i:27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>OR4F5_chr1_1_mid471</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>69546</td>\n",
       "      <td>60</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>LOC729737_chr1_1_mid4279</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>139030</td>\n",
       "      <td>60</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GGAGATGATGTCCTGGGACAGGAGTTGGCCGTGGAGAGGC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC729737</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>LOC100288069_chr1_7_mid303</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778566</td>\n",
       "      <td>48</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CACTCTGATGAGAAAACTGCCCAGCTCCAGGCACCATGGC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>LOC100288069_chr1_7_mid321</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778587</td>\n",
       "      <td>60</td>\n",
       "      <td>37M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CAGCTCCAGGCACCATGGCGCCCCAGTGATGTAGCCG</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:37</td>\n",
       "      <td>AS:i:37</td>\n",
       "      <td>XS:i:17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>LOC100288069_chr1_7_mid343</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778607</td>\n",
       "      <td>60</td>\n",
       "      <td>38M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCCCAGTGATGTAGCCGAACACCCGCGCCTCTAACGTC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:38</td>\n",
       "      <td>AS:i:38</td>\n",
       "      <td>XS:i:27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>LOC100288069_chr1_7_mid364</td>\n",
       "      <td>16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778632</td>\n",
       "      <td>60</td>\n",
       "      <td>37M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGCCTCTAACGTCGCCAACGGCCCACCTCTATGGTGT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:37</td>\n",
       "      <td>AS:i:37</td>\n",
       "      <td>XS:i:17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC100288069</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>FAM87B_chr1_1_mid511</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>817861</td>\n",
       "      <td>60</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACTGTGACTAAGGCTCAGTCCCTGAGGACTGCCTTGGCAT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>FAM87B_chr1_2_mid257</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>818960</td>\n",
       "      <td>60</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CACAGCATTAACTCTGGGAGTTGGGGACAGTGTGTATTCT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>FAM87B_chr1_2_mid279</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>818982</td>\n",
       "      <td>60</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GGGGACAGTGTGTATTCTTCCTCCAGACACTCTCTGTTTC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          QNAME FLAG RNAME     POS MAPQ CIGAR RNEXT  PNEXT  \\\n",
       "90         WASH7P_chr1_3_mid114   16  chr1   15891   60   39M     *    0.0   \n",
       "153         OR4F5_chr1_1_mid471    0  chr1   69546   60   40M     *    0.0   \n",
       "286    LOC729737_chr1_1_mid4279   16  chr1  139030   60   40M     *    0.0   \n",
       "689  LOC100288069_chr1_7_mid303   16  chr1  778566   48   40M     *    0.0   \n",
       "690  LOC100288069_chr1_7_mid321   16  chr1  778587   60   37M     *    0.0   \n",
       "691  LOC100288069_chr1_7_mid343   16  chr1  778607   60   38M     *    0.0   \n",
       "692  LOC100288069_chr1_7_mid364   16  chr1  778632   60   37M     *    0.0   \n",
       "707        FAM87B_chr1_1_mid511    0  chr1  817861   60   40M     *    0.0   \n",
       "716        FAM87B_chr1_2_mid257    0  chr1  818960   60   40M     *    0.0   \n",
       "717        FAM87B_chr1_2_mid279    0  chr1  818982   60   40M     *    0.0   \n",
       "\n",
       "     TLEN                                       SEQ QUAL      NM       MD  \\\n",
       "90    0.0   TCTAGCAGAGTGGCCAGCCACCGGAGGGGTCAACCACTT    *  NM:i:0  MD:Z:39   \n",
       "153   0.0  GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT    *  NM:i:0  MD:Z:40   \n",
       "286   0.0  GGAGATGATGTCCTGGGACAGGAGTTGGCCGTGGAGAGGC    *  NM:i:0  MD:Z:40   \n",
       "689   0.0  CACTCTGATGAGAAAACTGCCCAGCTCCAGGCACCATGGC    *  NM:i:0  MD:Z:40   \n",
       "690   0.0     CAGCTCCAGGCACCATGGCGCCCCAGTGATGTAGCCG    *  NM:i:0  MD:Z:37   \n",
       "691   0.0    CCCCAGTGATGTAGCCGAACACCCGCGCCTCTAACGTC    *  NM:i:0  MD:Z:38   \n",
       "692   0.0     CGCCTCTAACGTCGCCAACGGCCCACCTCTATGGTGT    *  NM:i:0  MD:Z:37   \n",
       "707   0.0  ACTGTGACTAAGGCTCAGTCCCTGAGGACTGCCTTGGCAT    *  NM:i:0  MD:Z:40   \n",
       "716   0.0  CACAGCATTAACTCTGGGAGTTGGGGACAGTGTGTATTCT    *  NM:i:0  MD:Z:40   \n",
       "717   0.0  GGGGACAGTGTGTATTCTTCCTCCAGACACTCTCTGTTTC    *  NM:i:0  MD:Z:40   \n",
       "\n",
       "          AS       XS   XA          gene   chr  \n",
       "90   AS:i:39  XS:i:27  NaN        WASH7P  chr1  \n",
       "153  AS:i:40  XS:i:20  NaN         OR4F5  chr1  \n",
       "286  AS:i:40  XS:i:25  NaN     LOC729737  chr1  \n",
       "689  AS:i:40  XS:i:32  NaN  LOC100288069  chr1  \n",
       "690  AS:i:37  XS:i:17  NaN  LOC100288069  chr1  \n",
       "691  AS:i:38  XS:i:27  NaN  LOC100288069  chr1  \n",
       "692  AS:i:37  XS:i:17  NaN  LOC100288069  chr1  \n",
       "707  AS:i:40  XS:i:30  NaN        FAM87B  chr1  \n",
       "716  AS:i:40  XS:i:30  NaN        FAM87B  chr1  \n",
       "717  AS:i:40  XS:i:18  NaN        FAM87B  chr1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QNAME</th>\n",
       "      <th>FLAG</th>\n",
       "      <th>RNAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>MAPQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>RNEXT</th>\n",
       "      <th>PNEXT</th>\n",
       "      <th>TLEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>NM</th>\n",
       "      <th>MD</th>\n",
       "      <th>AS</th>\n",
       "      <th>XS</th>\n",
       "      <th>XA</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1_chr1_1_mid18</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182395</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATG</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1_chr1_1_mid75</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11929</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCTTAGCCCAGACTTCCCGTGTCCTTTCCACCGGGCCTTT</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1_chr1_1_mid93</td>\n",
       "      <td>16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101978978</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ATCAAGACCCTGTGACCTCTCAAAGGCCCGGTGGAAAGGA</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1_chr1_1_mid125</td>\n",
       "      <td>16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101978948</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGTCAGACACCTGCAGATGAAGACCACAGCATCAAGACCC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1_chr1_1_mid148</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12</td>\n",
       "      <td>12132</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTG</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DDX11L1_chr1_1_mid180</td>\n",
       "      <td>16</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57214501</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GGAAAACTCCACTCCAGTGCTCAGCTTGCACCCTGGCACA</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DDX11L1_chr1_1_mid223</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "      <td>11760</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTC</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DDX11L1_chr1_1_mid251</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12</td>\n",
       "      <td>12236</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ATGGGCCATTGTTCATCTTCTGGCCCCTGTTGTCTGCATG</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DDX11L1_chr1_1_mid267</td>\n",
       "      <td>16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101978803</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTGGTTGTGGTATTAAGTTACATGCAGACAACAGGGGCCA</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DDX11L1_chr1_1_mid299</td>\n",
       "      <td>16</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101978773</td>\n",
       "      <td>0</td>\n",
       "      <td>40M</td>\n",
       "      <td>*</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTCATCTTTCCTCCAATCTTTCCCCTATGCCTGGTTGTGG</td>\n",
       "      <td>*</td>\n",
       "      <td>NM:i:0</td>\n",
       "      <td>MD:Z:40</td>\n",
       "      <td>AS:i:40</td>\n",
       "      <td>XS:i:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   QNAME FLAG  RNAME        POS MAPQ CIGAR RNEXT  PNEXT  TLEN  \\\n",
       "0   DDX11L1_chr1_1_mid18    0   chr1     182395    0   40M     *    0.0   0.0   \n",
       "1   DDX11L1_chr1_1_mid75    0   chr1      11929    0   40M     *    0.0   0.0   \n",
       "2   DDX11L1_chr1_1_mid93   16  chr15  101978978    0   40M     *    0.0   0.0   \n",
       "3  DDX11L1_chr1_1_mid125   16  chr15  101978948    0   40M     *    0.0   0.0   \n",
       "4  DDX11L1_chr1_1_mid148    0  chr12      12132    0   40M     *    0.0   0.0   \n",
       "5  DDX11L1_chr1_1_mid180   16   chrY   57214501    0   40M     *    0.0   0.0   \n",
       "6  DDX11L1_chr1_1_mid223    0  chr16      11760    0   40M     *    0.0   0.0   \n",
       "7  DDX11L1_chr1_1_mid251    0  chr12      12236    0   40M     *    0.0   0.0   \n",
       "8  DDX11L1_chr1_1_mid267   16  chr15  101978803    0   40M     *    0.0   0.0   \n",
       "9  DDX11L1_chr1_1_mid299   16  chr15  101978773    0   40M     *    0.0   0.0   \n",
       "\n",
       "                                        SEQ QUAL      NM       MD       AS  \\\n",
       "0  TGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATG    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "1  TCTTAGCCCAGACTTCCCGTGTCCTTTCCACCGGGCCTTT    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "2  ATCAAGACCCTGTGACCTCTCAAAGGCCCGGTGGAAAGGA    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "3  AGTCAGACACCTGCAGATGAAGACCACAGCATCAAGACCC    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "4  CATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTG    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "5  GGAAAACTCCACTCCAGTGCTCAGCTTGCACCCTGGCACA    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "6  TGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCATTGTTC    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "7  ATGGGCCATTGTTCATCTTCTGGCCCCTGTTGTCTGCATG    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "8  CTGGTTGTGGTATTAAGTTACATGCAGACAACAGGGGCCA    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "9  CTCATCTTTCCTCCAATCTTTCCCCTATGCCTGGTTGTGG    *  NM:i:0  MD:Z:40  AS:i:40   \n",
       "\n",
       "        XS   XA     gene   chr  \n",
       "0  XS:i:40  NaN  DDX11L1  chr1  \n",
       "1  XS:i:40  NaN  DDX11L1  chr1  \n",
       "2  XS:i:40  NaN  DDX11L1  chr1  \n",
       "3  XS:i:40  NaN  DDX11L1  chr1  \n",
       "4  XS:i:40  NaN  DDX11L1  chr1  \n",
       "5  XS:i:40  NaN  DDX11L1  chr1  \n",
       "6  XS:i:40  NaN  DDX11L1  chr1  \n",
       "7  XS:i:40  NaN  DDX11L1  chr1  \n",
       "8  XS:i:40  NaN  DDX11L1  chr1  \n",
       "9  XS:i:40  NaN  DDX11L1  chr1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mapq_thresh = 40\n",
    "samNames = \"bwamem_{{0}}.sam\".format() # formatted string that shows the name of the SAM files\n",
    "badprobes, uniqueprobes = {}, {}\n",
    "for chrm in chrms: # [14:15]\n",
    "    sam_out = pd.read_csv(os.path.join(savingdir, chrm, samNames.format(chrm)), sep = '\\t', names = ['QNAME', 'FLAG', 'RNAME', 'POS', \n",
    "                                           'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT',\n",
    "                                           'TLEN', 'SEQ', 'QUAL', \n",
    "                                          'NM', 'MD', 'AS', 'XS', 'XA'])\n",
    "    \n",
    "    sam_out = sam_out.loc[~sam_out['QNAME'].str.startswith('@')] # removing the headers\n",
    "    sam_out = pd.concat([sam_out, \n",
    "                         sam_out['QNAME'].str.extract(r\"(?P<gene>[0-9a-zA-Z-]+)_(?P<chr>\\S+)_\\d+_mid\\d+\")\n",
    "                        ], axis = 1) # extract the gene and chromosome names\n",
    "\n",
    "    # Removing probes with non-unique alignment\n",
    "    badprobes1 = sam_out.loc[sam_out['MAPQ'].astype(int) <= mapq_thresh ] \n",
    "    unqprbs = sam_out.loc[sam_out['MAPQ'].astype(int) > mapq_thresh] # removing probes with low mapping quality \n",
    "    badprobes2 = unqprbs.loc[~sam_out['XA'].isna()]\n",
    "    unqprbs = unqprbs.loc[sam_out['XA'].isna()] # removing probes with secondary alignment\n",
    "    badprobes3 = unqprbs.loc[sam_out['RNAME'].str.split(\"_\").str[0] != (sam_out['chr'])]\n",
    "    unqprbs =  unqprbs.loc[sam_out['RNAME'].str.split(\"_\").str[0] == (sam_out['chr'])]# removing probe whose chromosome doesn't match\n",
    "\n",
    "    badprobes[chrm] = pd.concat([badprobes1, badprobes2, badprobes3], ignore_index = True)\n",
    "    uniqueprobes[chrm] = unqprbs  \n",
    "    print(\"{3}: Removed {0} probes out of {1} ({2:.2f}%)\".format(badprobes[chrm].shape[0],\n",
    "                                                                badprobes[chrm].shape[0] + sam_out.shape[0],\n",
    "                                                                100*badprobes[chrm].shape[0] / (badprobes[chrm].shape[0] + sam_out.shape[0]),\n",
    "                                                                chrm))\n",
    "    \n",
    "display(uniqueprobes['chr1'].head(10))\n",
    "display(badprobes['chr1'].head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracting all info out of ppDesigner's output regarding the uniquely aligned probes\n",
    "goodprobes = {}\n",
    "for chrm in chrms:\n",
    "    ppd_out = pd.read_csv(os.path.join(savingdir, chrm, '{0}_0gap.ppd'.format(chrm)),\n",
    "                          sep = '\\t', header = None)\n",
    "    uniq_prbs = uniqueprobes[chrm]\n",
    "    ppd_out['full_name'] = [\"{0}_mid{1}\".format(row[0], row[3]) for _, row in ppd_out.iterrows()]\n",
    "    ppd_out = ppd_out.set_index('full_name')\n",
    "    good_prbs = ppd_out.loc[uniq_prbs['QNAME']]\n",
    "    goodprobes[chrm] = good_prbs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing the output to file\n",
    "Writing probes that didn't pass the filter as a SAM file with a \"badprobes\" suffix.\n",
    "\n",
    "Writing probes that did pass the filters in ppDesigner output format with \"filteredProbes\" prefix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the output to file\n",
    "for chrm in chrms:\n",
    "    badprobes[chrm].to_csv(os.path.join(savingdir, chrm, \n",
    "                                        'bwamem_{0}_badprobes.sam '.format(chrm)), sep = '\\t', index=False)\n",
    "    \n",
    "    goodprobes[chrm].to_csv(os.path.join(savingdir, chrm, \n",
    "                                         'filteredProbes_{0}.ppd'.format(chrm)), sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combining probes from all chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WASH7P_chr1_3</td>\n",
       "      <td>chr1:000015796-15947</td>\n",
       "      <td>0</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "      <td>AAGTGGTTGACCCCTCCGGT</td>\n",
       "      <td>64</td>\n",
       "      <td>GGCTGGCCACTCTGCTAGA</td>\n",
       "      <td>64</td>\n",
       "      <td>2.540034</td>\n",
       "      <td>C</td>\n",
       "      <td>AAGTGGTTGACCCCTCCGGTGGCTGGCCACTCTGCTAGA</td>\n",
       "      <td>1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F5_chr1_1</td>\n",
       "      <td>chr1:000069091-70008</td>\n",
       "      <td>0</td>\n",
       "      <td>471</td>\n",
       "      <td>471</td>\n",
       "      <td>GTTGGCGTTTGCCGTGC</td>\n",
       "      <td>64</td>\n",
       "      <td>ACTTACTCTTCTGTGGTCCCAAT</td>\n",
       "      <td>60</td>\n",
       "      <td>2.550166</td>\n",
       "      <td>W</td>\n",
       "      <td>GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LOC729737_chr1_1</td>\n",
       "      <td>chr1:000134773-139696</td>\n",
       "      <td>0</td>\n",
       "      <td>4279</td>\n",
       "      <td>4279</td>\n",
       "      <td>GCCTCTCCACGGCCAACT</td>\n",
       "      <td>63</td>\n",
       "      <td>CCTGTCCCAGGACATCATCTCC</td>\n",
       "      <td>64</td>\n",
       "      <td>2.548942</td>\n",
       "      <td>C</td>\n",
       "      <td>GCCTCTCCACGGCCAACTCCTGTCCCAGGACATCATCTCC</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LOC100288069_chr1_7</td>\n",
       "      <td>chr1:000778284-778688</td>\n",
       "      <td>0</td>\n",
       "      <td>303</td>\n",
       "      <td>303</td>\n",
       "      <td>GCCATGGTGCCTGGAGC</td>\n",
       "      <td>64</td>\n",
       "      <td>TGGGCAGTTTTCTCATCAGAGTG</td>\n",
       "      <td>62</td>\n",
       "      <td>2.542456</td>\n",
       "      <td>C</td>\n",
       "      <td>GCCATGGTGCCTGGAGCTGGGCAGTTTTCTCATCAGAGTG</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC100288069_chr1_7</td>\n",
       "      <td>chr1:000778284-778688</td>\n",
       "      <td>0</td>\n",
       "      <td>321</td>\n",
       "      <td>321</td>\n",
       "      <td>CGGCTACATCACTGGGGCG</td>\n",
       "      <td>63</td>\n",
       "      <td>CCATGGTGCCTGGAGCTG</td>\n",
       "      <td>63</td>\n",
       "      <td>1.239511</td>\n",
       "      <td>C</td>\n",
       "      <td>CGGCTACATCACTGGGGCGCCATGGTGCCTGGAGCTG</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529075</th>\n",
       "      <td>RBMY2FP_chrY_4</td>\n",
       "      <td>chrY:022314493-22316205</td>\n",
       "      <td>0</td>\n",
       "      <td>1429</td>\n",
       "      <td>1429</td>\n",
       "      <td>TTCAACAACACTCCACCA</td>\n",
       "      <td>57</td>\n",
       "      <td>TGTCTGAGAATGTGTGTATTCT</td>\n",
       "      <td>56</td>\n",
       "      <td>2.540504</td>\n",
       "      <td>W</td>\n",
       "      <td>TTCAACAACACTCCACCATGTCTGAGAATGTGTGTATTCT</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529076</th>\n",
       "      <td>LOC100652931_chrY_1</td>\n",
       "      <td>chrY:022316678-22316747</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>TCCTCCTCCTAAAAAGTCTGC</td>\n",
       "      <td>60</td>\n",
       "      <td>TCCTTCTGCTGTGGTAAGA</td>\n",
       "      <td>56</td>\n",
       "      <td>2.546745</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTCCTCCTAAAAAGTCTGCTCCTTCTGCTGTGGTAAGA</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529077</th>\n",
       "      <td>LOC100652931_chrY_1</td>\n",
       "      <td>chrY:022316678-22316747</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>TGCTCCTTCTGCTGTGGT</td>\n",
       "      <td>61</td>\n",
       "      <td>AAGAAGCAATAATTGGATGAGA</td>\n",
       "      <td>55</td>\n",
       "      <td>2.592248</td>\n",
       "      <td>W</td>\n",
       "      <td>TGCTCCTTCTGCTGTGGTAAGAAGCAATAATTGGATGAGA</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529078</th>\n",
       "      <td>LOC100652931_chrY_2</td>\n",
       "      <td>chrY:022318783-22318897</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "      <td>TCCTACACGTAGACGGCCA</td>\n",
       "      <td>63</td>\n",
       "      <td>ATCTCTTCCTGGAGAAATGAC</td>\n",
       "      <td>57</td>\n",
       "      <td>2.550421</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTACACGTAGACGGCCAATCTCTTCCTGGAGAAATGAC</td>\n",
       "      <td>1.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529079</th>\n",
       "      <td>LOC100652931_chrY_2</td>\n",
       "      <td>chrY:022318783-22318897</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>77</td>\n",
       "      <td>TCCTGGAGAAATGACCACA</td>\n",
       "      <td>58</td>\n",
       "      <td>TGTCTCTAAGAGACGATGGTT</td>\n",
       "      <td>57</td>\n",
       "      <td>2.532303</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTGGAGAAATGACCACATGTCTCTAAGAGACGATGGTT</td>\n",
       "      <td>1.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1529080 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0                        1  2     3     4  \\\n",
       "0              WASH7P_chr1_3     chr1:000015796-15947  0   114   114   \n",
       "1               OR4F5_chr1_1     chr1:000069091-70008  0   471   471   \n",
       "2           LOC729737_chr1_1    chr1:000134773-139696  0  4279  4279   \n",
       "3        LOC100288069_chr1_7    chr1:000778284-778688  0   303   303   \n",
       "4        LOC100288069_chr1_7    chr1:000778284-778688  0   321   321   \n",
       "...                      ...                      ... ..   ...   ...   \n",
       "1529075       RBMY2FP_chrY_4  chrY:022314493-22316205  0  1429  1429   \n",
       "1529076  LOC100652931_chrY_1  chrY:022316678-22316747  0    25    25   \n",
       "1529077  LOC100652931_chrY_1  chrY:022316678-22316747  0    40    40   \n",
       "1529078  LOC100652931_chrY_2  chrY:022318783-22318897  0    52    52   \n",
       "1529079  LOC100652931_chrY_2  chrY:022318783-22318897  0    77    77   \n",
       "\n",
       "                             5   6                        7   8         9 10  \\\n",
       "0         AAGTGGTTGACCCCTCCGGT  64      GGCTGGCCACTCTGCTAGA  64  2.540034  C   \n",
       "1            GTTGGCGTTTGCCGTGC  64  ACTTACTCTTCTGTGGTCCCAAT  60  2.550166  W   \n",
       "2           GCCTCTCCACGGCCAACT  63   CCTGTCCCAGGACATCATCTCC  64  2.548942  C   \n",
       "3            GCCATGGTGCCTGGAGC  64  TGGGCAGTTTTCTCATCAGAGTG  62  2.542456  C   \n",
       "4          CGGCTACATCACTGGGGCG  63       CCATGGTGCCTGGAGCTG  63  1.239511  C   \n",
       "...                        ...  ..                      ...  ..       ... ..   \n",
       "1529075     TTCAACAACACTCCACCA  57   TGTCTGAGAATGTGTGTATTCT  56  2.540504  W   \n",
       "1529076  TCCTCCTCCTAAAAAGTCTGC  60      TCCTTCTGCTGTGGTAAGA  56  2.546745  W   \n",
       "1529077     TGCTCCTTCTGCTGTGGT  61   AAGAAGCAATAATTGGATGAGA  55  2.592248  W   \n",
       "1529078    TCCTACACGTAGACGGCCA  63    ATCTCTTCCTGGAGAAATGAC  57  2.550421  W   \n",
       "1529079    TCCTGGAGAAATGACCACA  58    TGTCTCTAAGAGACGATGGTT  57  2.532303  W   \n",
       "\n",
       "                                               11   12  \n",
       "0         AAGTGGTTGACCCCTCCGGTGGCTGGCCACTCTGCTAGA  1.3  \n",
       "1        GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT  0.2  \n",
       "2        GCCTCTCCACGGCCAACTCCTGTCCCAGGACATCATCTCC  0.0  \n",
       "3        GCCATGGTGCCTGGAGCTGGGCAGTTTTCTCATCAGAGTG  0.5  \n",
       "4           CGGCTACATCACTGGGGCGCCATGGTGCCTGGAGCTG  0.5  \n",
       "...                                           ...  ...  \n",
       "1529075  TTCAACAACACTCCACCATGTCTGAGAATGTGTGTATTCT  0.1  \n",
       "1529076  TCCTCCTCCTAAAAAGTCTGCTCCTTCTGCTGTGGTAAGA  2.9  \n",
       "1529077  TGCTCCTTCTGCTGTGGTAAGAAGCAATAATTGGATGAGA  2.9  \n",
       "1529078  TCCTACACGTAGACGGCCAATCTCTTCCTGGAGAAATGAC  1.7  \n",
       "1529079  TCCTGGAGAAATGACCACATGTCTCTAAGAGACGATGGTT  1.7  \n",
       "\n",
       "[1529080 rows x 13 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np, os, pandas as pd\n",
    "savingdir = os.getcwd()\n",
    "\n",
    "chrms = ['chr{}'.format(n) for n in [*np.arange(1, 23), 'X', 'Y']]\n",
    "\n",
    "allprobes = []\n",
    "for chrm in chrms:\n",
    "    allprobes.append(pd.read_csv(os.path.join(savingdir, chrm, \n",
    "                                         'filteredProbes_{0}.ppd'.format(chrm)),\n",
    "                                sep = '\\t'))\n",
    "allprobes = pd.concat(allprobes, ignore_index=True)\n",
    "allprobes                     \n",
    "                     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing probes with DpnII motif\n",
    "removing all probes that have the DpnII motif in either of their arms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "132731 probes were filtered due to DpnII motif filtering\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WASH7P_chr1_3</td>\n",
       "      <td>chr1:000015796-15947</td>\n",
       "      <td>0</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "      <td>AAGTGGTTGACCCCTCCGGT</td>\n",
       "      <td>64</td>\n",
       "      <td>GGCTGGCCACTCTGCTAGA</td>\n",
       "      <td>64</td>\n",
       "      <td>2.540034</td>\n",
       "      <td>C</td>\n",
       "      <td>AAGTGGTTGACCCCTCCGGTGGCTGGCCACTCTGCTAGA</td>\n",
       "      <td>1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F5_chr1_1</td>\n",
       "      <td>chr1:000069091-70008</td>\n",
       "      <td>0</td>\n",
       "      <td>471</td>\n",
       "      <td>471</td>\n",
       "      <td>GTTGGCGTTTGCCGTGC</td>\n",
       "      <td>64</td>\n",
       "      <td>ACTTACTCTTCTGTGGTCCCAAT</td>\n",
       "      <td>60</td>\n",
       "      <td>2.550166</td>\n",
       "      <td>W</td>\n",
       "      <td>GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LOC729737_chr1_1</td>\n",
       "      <td>chr1:000134773-139696</td>\n",
       "      <td>0</td>\n",
       "      <td>4279</td>\n",
       "      <td>4279</td>\n",
       "      <td>GCCTCTCCACGGCCAACT</td>\n",
       "      <td>63</td>\n",
       "      <td>CCTGTCCCAGGACATCATCTCC</td>\n",
       "      <td>64</td>\n",
       "      <td>2.548942</td>\n",
       "      <td>C</td>\n",
       "      <td>GCCTCTCCACGGCCAACTCCTGTCCCAGGACATCATCTCC</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LOC100288069_chr1_7</td>\n",
       "      <td>chr1:000778284-778688</td>\n",
       "      <td>0</td>\n",
       "      <td>303</td>\n",
       "      <td>303</td>\n",
       "      <td>GCCATGGTGCCTGGAGC</td>\n",
       "      <td>64</td>\n",
       "      <td>TGGGCAGTTTTCTCATCAGAGTG</td>\n",
       "      <td>62</td>\n",
       "      <td>2.542456</td>\n",
       "      <td>C</td>\n",
       "      <td>GCCATGGTGCCTGGAGCTGGGCAGTTTTCTCATCAGAGTG</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC100288069_chr1_7</td>\n",
       "      <td>chr1:000778284-778688</td>\n",
       "      <td>0</td>\n",
       "      <td>321</td>\n",
       "      <td>321</td>\n",
       "      <td>CGGCTACATCACTGGGGCG</td>\n",
       "      <td>63</td>\n",
       "      <td>CCATGGTGCCTGGAGCTG</td>\n",
       "      <td>63</td>\n",
       "      <td>1.239511</td>\n",
       "      <td>C</td>\n",
       "      <td>CGGCTACATCACTGGGGCGCCATGGTGCCTGGAGCTG</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529075</th>\n",
       "      <td>RBMY2FP_chrY_4</td>\n",
       "      <td>chrY:022314493-22316205</td>\n",
       "      <td>0</td>\n",
       "      <td>1429</td>\n",
       "      <td>1429</td>\n",
       "      <td>TTCAACAACACTCCACCA</td>\n",
       "      <td>57</td>\n",
       "      <td>TGTCTGAGAATGTGTGTATTCT</td>\n",
       "      <td>56</td>\n",
       "      <td>2.540504</td>\n",
       "      <td>W</td>\n",
       "      <td>TTCAACAACACTCCACCATGTCTGAGAATGTGTGTATTCT</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529076</th>\n",
       "      <td>LOC100652931_chrY_1</td>\n",
       "      <td>chrY:022316678-22316747</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>TCCTCCTCCTAAAAAGTCTGC</td>\n",
       "      <td>60</td>\n",
       "      <td>TCCTTCTGCTGTGGTAAGA</td>\n",
       "      <td>56</td>\n",
       "      <td>2.546745</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTCCTCCTAAAAAGTCTGCTCCTTCTGCTGTGGTAAGA</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529077</th>\n",
       "      <td>LOC100652931_chrY_1</td>\n",
       "      <td>chrY:022316678-22316747</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>TGCTCCTTCTGCTGTGGT</td>\n",
       "      <td>61</td>\n",
       "      <td>AAGAAGCAATAATTGGATGAGA</td>\n",
       "      <td>55</td>\n",
       "      <td>2.592248</td>\n",
       "      <td>W</td>\n",
       "      <td>TGCTCCTTCTGCTGTGGTAAGAAGCAATAATTGGATGAGA</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529078</th>\n",
       "      <td>LOC100652931_chrY_2</td>\n",
       "      <td>chrY:022318783-22318897</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "      <td>TCCTACACGTAGACGGCCA</td>\n",
       "      <td>63</td>\n",
       "      <td>ATCTCTTCCTGGAGAAATGAC</td>\n",
       "      <td>57</td>\n",
       "      <td>2.550421</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTACACGTAGACGGCCAATCTCTTCCTGGAGAAATGAC</td>\n",
       "      <td>1.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529079</th>\n",
       "      <td>LOC100652931_chrY_2</td>\n",
       "      <td>chrY:022318783-22318897</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>77</td>\n",
       "      <td>TCCTGGAGAAATGACCACA</td>\n",
       "      <td>58</td>\n",
       "      <td>TGTCTCTAAGAGACGATGGTT</td>\n",
       "      <td>57</td>\n",
       "      <td>2.532303</td>\n",
       "      <td>W</td>\n",
       "      <td>TCCTGGAGAAATGACCACATGTCTCTAAGAGACGATGGTT</td>\n",
       "      <td>1.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1396349 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0                        1  2     3     4  \\\n",
       "0              WASH7P_chr1_3     chr1:000015796-15947  0   114   114   \n",
       "1               OR4F5_chr1_1     chr1:000069091-70008  0   471   471   \n",
       "2           LOC729737_chr1_1    chr1:000134773-139696  0  4279  4279   \n",
       "3        LOC100288069_chr1_7    chr1:000778284-778688  0   303   303   \n",
       "4        LOC100288069_chr1_7    chr1:000778284-778688  0   321   321   \n",
       "...                      ...                      ... ..   ...   ...   \n",
       "1529075       RBMY2FP_chrY_4  chrY:022314493-22316205  0  1429  1429   \n",
       "1529076  LOC100652931_chrY_1  chrY:022316678-22316747  0    25    25   \n",
       "1529077  LOC100652931_chrY_1  chrY:022316678-22316747  0    40    40   \n",
       "1529078  LOC100652931_chrY_2  chrY:022318783-22318897  0    52    52   \n",
       "1529079  LOC100652931_chrY_2  chrY:022318783-22318897  0    77    77   \n",
       "\n",
       "                             5   6                        7   8         9 10  \\\n",
       "0         AAGTGGTTGACCCCTCCGGT  64      GGCTGGCCACTCTGCTAGA  64  2.540034  C   \n",
       "1            GTTGGCGTTTGCCGTGC  64  ACTTACTCTTCTGTGGTCCCAAT  60  2.550166  W   \n",
       "2           GCCTCTCCACGGCCAACT  63   CCTGTCCCAGGACATCATCTCC  64  2.548942  C   \n",
       "3            GCCATGGTGCCTGGAGC  64  TGGGCAGTTTTCTCATCAGAGTG  62  2.542456  C   \n",
       "4          CGGCTACATCACTGGGGCG  63       CCATGGTGCCTGGAGCTG  63  1.239511  C   \n",
       "...                        ...  ..                      ...  ..       ... ..   \n",
       "1529075     TTCAACAACACTCCACCA  57   TGTCTGAGAATGTGTGTATTCT  56  2.540504  W   \n",
       "1529076  TCCTCCTCCTAAAAAGTCTGC  60      TCCTTCTGCTGTGGTAAGA  56  2.546745  W   \n",
       "1529077     TGCTCCTTCTGCTGTGGT  61   AAGAAGCAATAATTGGATGAGA  55  2.592248  W   \n",
       "1529078    TCCTACACGTAGACGGCCA  63    ATCTCTTCCTGGAGAAATGAC  57  2.550421  W   \n",
       "1529079    TCCTGGAGAAATGACCACA  58    TGTCTCTAAGAGACGATGGTT  57  2.532303  W   \n",
       "\n",
       "                                               11   12  \n",
       "0         AAGTGGTTGACCCCTCCGGTGGCTGGCCACTCTGCTAGA  1.3  \n",
       "1        GTTGGCGTTTGCCGTGCACTTACTCTTCTGTGGTCCCAAT  0.2  \n",
       "2        GCCTCTCCACGGCCAACTCCTGTCCCAGGACATCATCTCC  0.0  \n",
       "3        GCCATGGTGCCTGGAGCTGGGCAGTTTTCTCATCAGAGTG  0.5  \n",
       "4           CGGCTACATCACTGGGGCGCCATGGTGCCTGGAGCTG  0.5  \n",
       "...                                           ...  ...  \n",
       "1529075  TTCAACAACACTCCACCATGTCTGAGAATGTGTGTATTCT  0.1  \n",
       "1529076  TCCTCCTCCTAAAAAGTCTGCTCCTTCTGCTGTGGTAAGA  2.9  \n",
       "1529077  TGCTCCTTCTGCTGTGGTAAGAAGCAATAATTGGATGAGA  2.9  \n",
       "1529078  TCCTACACGTAGACGGCCAATCTCTTCCTGGAGAAATGAC  1.7  \n",
       "1529079  TCCTGGAGAAATGACCACATGTCTCTAAGAGACGATGGTT  1.7  \n",
       "\n",
       "[1396349 rows x 13 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dpnii_pat = 'GATC'\n",
    "probes_dpn2free = allprobes.loc[~(allprobes['5'].str.contains(dpnii_pat) | allprobes['7'].str.contains(dpnii_pat))]\n",
    "print(\"{} probes were filtered due to DpnII motif filtering\".format(allprobes.shape[0] - probes_dpn2free.shape[0]))\n",
    "probes_dpn2free"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the final probe list in the .ppd format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_dpn2free.to_csv(os.path.join(savingdir, \"probes_DpnII-filtered.ppd\"), \n",
    "                      sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting the number of probes per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name  chr  \n",
       "A1BG       chr19     57\n",
       "A1BG-AS1   chr19     44\n",
       "A1CF       chr10    129\n",
       "A2M        chr12    108\n",
       "A2M-AS1    chr12     24\n",
       "                   ... \n",
       "ZYG11A     chr1      88\n",
       "ZYG11B     chr1     117\n",
       "ZYX        chr7      43\n",
       "ZZEF1      chr17    272\n",
       "ZZZ3       chr1     105\n",
       "Length: 26601, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_probes = probes_dpn2free.copy()\n",
    "all_probes['gene_name'] = all_probes['0'].str.split('_').str[0]\n",
    "all_probes['chr'] = all_probes['1'].str.split(':').str[0].to_numpy()\n",
    "gene_counts = all_probes.groupby(by = ['gene_name', 'chr']).size()\n",
    "gene_counts.to_csv(os.path.join(savingdir, 'probe_counts_DpnII.tsv'), \n",
    "                   sep = '\\t', header = ['probe_count'])\n",
    "gene_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make a bed file for the probes to visualize in UCSC Genome Browser\n",
    "If using Matt's verion of ppDesigner, a correction of $flankingLen needs to be applied to the probe locations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "flankingLen = 0\n",
    "all_probes = probes_dpn2free.copy()\n",
    "all_probes['chr'] = all_probes['1'].str.split(':').str[0].to_numpy()\n",
    "all_probes['gene_name'] = all_probes['0'].str.split('_').str[0]\n",
    "\n",
    "\n",
    "exon_start = all_probes['1'].str.split(':').str[1].str.split('-').str[0].astype(int)\n",
    "leftArmLen = all_probes['5'].str.len()\n",
    "rightArmLen = all_probes['7'].str.len()\n",
    "\n",
    "all_probes['probe_left'] = exon_start\n",
    "all_probes['probe_right'] = exon_start\n",
    "all_probes.loc[all_probes['10'] == 'W', 'probe_left'] += (all_probes['3'] - leftArmLen - flankingLen)\n",
    "all_probes.loc[all_probes['10'] == 'C', 'probe_left'] += (all_probes['3']- rightArmLen - flankingLen)\n",
    "all_probes.loc[all_probes['10'] == 'W', 'probe_right'] += (all_probes['3'] + rightArmLen - flankingLen - 1)\n",
    "all_probes.loc[all_probes['10'] == 'C', 'probe_right'] += (all_probes['3'] + leftArmLen - flankingLen - 1)\n",
    "\n",
    "probe_ranks = all_probes.groupby('gene_name').cumcount() + 1\n",
    "all_probes['probe_name'] = all_probes['gene_name'] + '_' + probe_ranks.astype(str)\n",
    "bed = pd.DataFrame({'chr' : all_probes['chr'], 'left' : all_probes['probe_left'].astype(int), \n",
    "                    'right' : all_probes['probe_right'].astype(int), \n",
    "                    'name' : all_probes['probe_name']}, index = None)\n",
    "bed.to_csv(os.path.join(savingdir, 'probe_ranges_DpnII.bed'), \n",
    "           sep = '\\t', index = False, \n",
    "           header = ['track', 'name=ppD_DF_DpnII_refseq_overlap20', \n",
    "                    \"description='Probes designed by 221008 version of ppDesigner (max_overlap 20) using Refseq annotation. DpnII containing probes filtered.'\", \n",
    "                    'useScore=0'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "### wrong code for generating bed files. Messes up for genes on the negative strand\n",
    "# prb_left = exon_start + all_probes['3'].astype(int) - leftArmLen - flankingLen + 1\n",
    "# prb_right= exon_start + all_probes['3'].astype(int) + rightArmLen - flankingLen + 1\n",
    "\n",
    "# probe_ranks = all_probes.groupby('gene_name').cumcount() + 1\n",
    "# all_probes['probe_name'] = all_probes['gene_name'] + '_' + probe_ranks.astype(str)\n",
    "# bed = pd.DataFrame({'chr' : all_probes['chr'], 'leftArm' : prb_left, 'rightArm' : prb_right, \n",
    "#                   'name' : all_probes['probe_name']}, index = None)\n",
    "# bed.to_csv(os.path.join(savingdir, 'probe_ranges_DpnII_largeExon{}_3.bed'.format(largeExon)), \n",
    "#            sep = '\\t', index = False, \n",
    "#            header = ['track', 'name=MattPPD_DpnII_refseq_{}'.format(largeExon), \n",
    "#                     \"description='Probes designed by Matts version of ppDesigner (largeExon {0}) using Refseq annotation. DpnII containing probes filtered.'\".format(largeExon), \n",
    "#                     'useScore=0'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Miscellaneous codes, old codes, scratch codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Counting the number of probes per gene\n",
    "# counts = []\n",
    "# for chrm in chrms:\n",
    "# #     counts[chrm] = {}\n",
    "#     probes_chr = pd.read_csv(os.path.join(savingdir, chrm, \n",
    "#                                          'filteredProbes_{0}_largeExon{1}.ppd'.format(chrm, largeExon)), \n",
    "#                              sep = '\\t')\n",
    "#     probes_chr['gene_name'] = probes_chr['0'].str.split('_').str[0]\n",
    "# #     counts[chrm] = probes_chr.groupby(by = ['gene_name']).size().to_dict()\n",
    "#     gene_chr = probes_chr.groupby(by = ['gene_name']).size()\n",
    "#     count_chr = pd.DataFrame({'chr' : gene_chr.shape[0] * [chrm],\n",
    "#                               'probe_count' : gene_chr})\n",
    "#     counts.append(count_chr) \n",
    "# counts = pd.concat(counts)\n",
    "# counts.to_csv(os.path.join(savingdir, 'probe_counts_largeExon{}.tsv'.format(largeExon)), sep = '\\t')\n",
    "# counts.head()\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_file = os.path.join('/media/Scratch_SSD_Voyager/Kian/genomes/hg38/hg38.ncbiRefSeq.gtf')\n",
    "refseq_df = gtfparse.read_gtf(refseq_file)\n",
    "refseq_df = refseq_df.loc[(refseq_df['feature'] == 'exon') & (refseq_df['transcript_id'].str.startswith('N'))]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_df.loc[refseq_df['gene_name'] == 'B2M'].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\", 201)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.get_option(\"display.max_rows\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.get_option(\"display.min_rows\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.min_rows\", 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# samNames = \"bwamem_{{0}}_largeExon{0}.sam\".format(largeExon) # formatted string that shows the name of the SAM files\n",
    "# badprobes, uniqprobes = {}, {}\n",
    "# for chrm in chrms:\n",
    "#     sam_out = pd.read_csv(os.path.join(savingdir, chrm, samNames.format(chrm)), sep = '\\t', names = ['QNAME', 'FLAG', 'RNAME', 'POS', \n",
    "#                                            'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT',\n",
    "#                                            'TLEN', 'SEQ', 'QUAL', \n",
    "#                                           'NM', 'MD', 'AS', 'XS', 'XA'])\n",
    "    \n",
    "#     # Removing probes with non-unique alignment\n",
    "#     sam_out = sam_out.loc[~sam_out['QNAME'].str.startswith('@')] # removing the headers\n",
    "#     badprobes1 = sam_out.loc[sam_out['MAPQ'].astype(int) <= 20 ] \n",
    "#     sam_out = sam_out.loc[sam_out['MAPQ'].astype(int) > 20] # removing probes with mapping quality of zero\n",
    "#     badprobes2 = sam_out.loc[~sam_out['XA'].isna()]\n",
    "#     sam_out = sam_out.loc[sam_out['XA'].isna()] # removing probes with secondary alignment\n",
    "#     badprobes3 = sam_out.loc[~sam_out['RNAME'].str.startswith(chrm)]\n",
    "#     sam_out = sam_out.loc[sam_out['RNAME'].str.startswith(chrm)] # removing probe whose chromosome doesn't match\n",
    "\n",
    "#     badprobes[chrm] = pd.concat([badprobes1, badprobes2, badprobes3], ignore_index = True)\n",
    "#     uniqprobes[chrm] = sam_out  \n",
    "#     print(\"{3}: Removed {0} probes out of {1} ({2:.2f}%)\".format(badprobes[chrm].shape[0],\n",
    "#                                                                 badprobes[chrm].shape[0] + sam_out.shape[0],\n",
    "#                                                                 100*badprobes[chrm].shape[0] / (badprobes[chrm].shape[0] + sam_out.shape[0]),\n",
    "#                                                                 chrm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\"Sam_out shape: {}\".format(sam_out.shape))\n",
    "# print(\"Alt_sam_out shape: {}\".format(alt_sam_out.shape))\n",
    "# print(\"norm_sam_out shape: {}\".format(norm_sam_out.shape))\n",
    "# print(\"bad_probes shape: {}\".format(badprobes['chr15'].shape))\n",
    "# print(\"alt_bad_probes shape: {}\".format(alt_badprobes.shape))\n",
    "# print(\"norm_bad_probes shape: {}\".format(norm_badprobes.shape))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
